// kepler.gl/dist/components/side-panel/layer-manager.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp2;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  .ui-sortable {\n    display: block;\n    position: relative;\n    overflow: visible;\n    user-select: none;\n\n    :before {\n      content: \' \';\n      display: table;\n    }\n\n    :after {\n      content: \' \';\n      display: table;\n    }\n  }\n\n  .ui-sortable-item.ui-sortable-dragging {\n    position: absolute;\n    z-index: 1688;\n    cursor: move;\n  }\n\n  .ui-sortable-item.ui-sortable-dragging:hover {\n    cursor: move;\n    opacity: 0.5;\n  }\n\n  .ui-sortable-placeholder {\n    display: none;\n  }\n\n  .ui-sortable-placeholder.visible {\n    display: block;\n    opacity: 0;\n    z-index: -1;\n  }\n'], ['\n  .ui-sortable {\n    display: block;\n    position: relative;\n    overflow: visible;\n    user-select: none;\n\n    :before {\n      content: \' \';\n      display: table;\n    }\n\n    :after {\n      content: \' \';\n      display: table;\n    }\n  }\n\n  .ui-sortable-item.ui-sortable-dragging {\n    position: absolute;\n    z-index: 1688;\n    cursor: move;\n  }\n\n  .ui-sortable-item.ui-sortable-dragging:hover {\n    cursor: move;\n    opacity: 0.5;\n  }\n\n  .ui-sortable-placeholder {\n    display: none;\n  }\n\n  .ui-sortable-placeholder.visible {\n    display: block;\n    opacity: 0;\n    z-index: -1;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reactAnythingSortable = require('react-anything-sortable');

var _reactAnythingSortable2 = _interopRequireDefault(_reactAnythingSortable);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _reselect = require('reselect');

var _layerPanel = require('./layer-panel/layer-panel');

var _layerPanel2 = _interopRequireDefault(_layerPanel);

var _sourceDataCatalog = require('./source-data-catalog');

var _sourceDataCatalog2 = _interopRequireDefault(_sourceDataCatalog);

var _icons = require('../common/icons');

var _itemSelector = require('../common/item-selector/item-selector');

var _itemSelector2 = _interopRequireDefault(_itemSelector);

var _styledComponents3 = require('../common/styled-components');

var _defaultSettings = require('../../constants/default-settings');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StyledSortable = _styledComponents2.default.div(_templateObject);

var LayerManager = (_temp2 = _class = function (_Component) {
  (0, _inherits3.default)(LayerManager, _Component);

  function LayerManager() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, LayerManager);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = LayerManager.__proto__ || Object.getPrototypeOf(LayerManager)).call.apply(_ref, [this].concat(args))), _this), _this.layerClassSelector = function (props) {
      return props.layerClasses;
    }, _this.layerTypeOptionsSelector = (0, _reselect.createSelector)(_this.layerClassSelector, function (layerClasses) {
      return Object.keys(layerClasses).map(function (key) {
        var layer = new layerClasses[key]();
        return {
          id: key,
          label: layer.name,
          icon: layer.layerIcon
        };
      });
    }), _this._addEmptyNewLayer = function () {
      _this.props.addLayer();
    }, _this._handleSort = function (order) {
      _this.props.updateLayerOrder(order);
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(LayerManager, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          layers = _props.layers,
          datasets = _props.datasets,
          layerOrder = _props.layerOrder,
          openModal = _props.openModal;

      var defaultDataset = Object.keys(datasets)[0];
      var layerTypeOptions = this.layerTypeOptionsSelector(this.props);

      var layerActions = {
        layerConfigChange: this.props.layerConfigChange,
        layerVisualChannelConfigChange: this.props.layerVisualChannelConfigChange,
        layerTypeChange: this.props.layerTypeChange,
        layerVisConfigChange: this.props.layerVisConfigChange,
        removeLayer: this.props.removeLayer
      };

      var panelProps = { datasets: datasets, openModal: openModal, layerTypeOptions: layerTypeOptions };

      return _react2.default.createElement(
        StyledSortable,
        { className: 'layer-manager' },
        _react2.default.createElement(_sourceDataCatalog2.default, {
          datasets: datasets,
          showDatasetTable: this.props.showDatasetTable,
          removeDataset: this.props.removeDataset,
          showDeleteDataset: true
        }),
        _react2.default.createElement(
          _styledComponents3.Button,
          {
            onClick: this.props.showAddDataModal,
            isInactive: !defaultDataset,
            width: '105px',
            secondary: true
          },
          _react2.default.createElement(_icons.Add, { height: '12px' }),
          'Add Data'
        ),
        _react2.default.createElement(_styledComponents3.SidePanelDivider, null),
        _react2.default.createElement(
          _styledComponents3.SidePanelSection,
          null,
          _react2.default.createElement(
            _reactAnythingSortable2.default,
            {
              onSort: this._handleSort,
              direction: 'vertical',
              sortHandle: 'sort--handle',
              dynamic: true
            },
            layerOrder.map(function (idx) {
              return _react2.default.createElement(_layerPanel2.default, (0, _extends3.default)({}, panelProps, layerActions, {
                sortData: idx,
                key: layers[idx].id,
                idx: idx,
                layer: layers[idx]
              }));
            })
          )
        ),
        _react2.default.createElement(
          _styledComponents3.SidePanelSection,
          null,
          defaultDataset ? _react2.default.createElement(
            _styledComponents3.Button,
            { onClick: this._addEmptyNewLayer, width: '105px' },
            _react2.default.createElement(_icons.Add, { height: '12px' }),
            'Add Layer'
          ) : null
        ),
        _react2.default.createElement(LayerBlendingSelector, {
          layerBlending: this.props.layerBlending,
          updateLayerBlending: this.props.updateLayerBlending
        })
      );
    }
  }]);
  return LayerManager;
}(_react.Component), _class.propTypes = {
  addLayer: _propTypes2.default.func.isRequired,
  datasets: _propTypes2.default.object.isRequired,
  layerBlending: _propTypes2.default.string.isRequired,
  layerClasses: _propTypes2.default.object.isRequired,
  layers: _propTypes2.default.arrayOf(_propTypes2.default.any).isRequired,
  layerConfigChange: _propTypes2.default.func.isRequired,
  layerVisualChannelConfigChange: _propTypes2.default.func.isRequired,
  layerTypeChange: _propTypes2.default.func.isRequired,
  layerVisConfigChange: _propTypes2.default.func.isRequired,
  openModal: _propTypes2.default.func.isRequired,
  removeLayer: _propTypes2.default.func.isRequired,
  removeDataset: _propTypes2.default.func.isRequired,
  showDatasetTable: _propTypes2.default.func.isRequired,
  updateLayerBlending: _propTypes2.default.func.isRequired,
  updateLayerOrder: _propTypes2.default.func.isRequired
}, _temp2);
exports.default = LayerManager;


var LayerBlendingSelector = function LayerBlendingSelector(_ref2) {
  var layerBlending = _ref2.layerBlending,
      updateLayerBlending = _ref2.updateLayerBlending;
  return _react2.default.createElement(
    _styledComponents3.SidePanelSection,
    null,
    _react2.default.createElement(
      _styledComponents3.PanelLabel,
      null,
      'Layer Blending'
    ),
    _react2.default.createElement(_itemSelector2.default, {
      selectedItems: layerBlending,
      options: Object.keys(_defaultSettings.LAYER_BLENDINGS),
      multiSelect: false,
      searchable: false,
      onChange: updateLayerBlending
    })
  );
};
