// kepler.gl/dist/components/side-panel/panel-toggle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  background-color: ', ';\n  padding: 0 16px;\n  display: flex;\n  min-height: 30px;\n'], ['\n  background-color: ', ';\n  padding: 0 16px;\n  display: flex;\n  min-height: 30px;\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  align-items: flex-end;\n  border-bottom-style: solid;\n  border-bottom-width: 2px;\n  border-bottom-color: ', ';\n  color: ', ';\n  display: flex;\n  justify-content: center;\n  margin-right: 12px;\n  padding-bottom: 6px;\n  width: 30px;\n  \n  :hover {\n    cursor: pointer;\n    color: ', ';\n  }\n'], ['\n  align-items: flex-end;\n  border-bottom-style: solid;\n  border-bottom-width: 2px;\n  border-bottom-color: ', ';\n  color: ', ';\n  display: flex;\n  justify-content: center;\n  margin-right: 12px;\n  padding-bottom: 6px;\n  width: 30px;\n  \n  :hover {\n    cursor: pointer;\n    color: ', ';\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _styledComponents3 = require('../common/styled-components');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  panels: _propTypes2.default.array,
  activePanel: _propTypes2.default.string,
  togglePanel: _propTypes2.default.func
};

var PanelHeaderBottom = _styledComponents2.default.div.attrs({
  className: 'side-side-panel__header__bottom'
})(_templateObject, function (props) {
  return props.theme.sidePanelHeaderBg;
});

var PanelTab = _styledComponents2.default.div.attrs({
  className: 'side-panel__tab'
})(_templateObject2, function (props) {
  return props.active ? props.theme.subtextColorActive : 'transparent';
}, function (props) {
  return props.active ? props.theme.subtextColorActive : props.theme.subtextColor;
}, function (props) {
  return props.theme.textColorHl;
});

var PanelToggle = function PanelToggle(_ref) {
  var panels = _ref.panels,
      activePanel = _ref.activePanel,
      togglePanel = _ref.togglePanel;
  return _react2.default.createElement(
    PanelHeaderBottom,
    null,
    panels.map(function (panel) {
      return _react2.default.createElement(
        PanelTab,
        {
          key: panel.id,
          'data-tip': true,
          'data-for': panel.id + '-nav',
          active: activePanel === panel.id,
          onClick: function onClick() {
            return togglePanel(panel.id);
          }
        },
        _react2.default.createElement(panel.iconComponent, { height: '20px' }),
        _react2.default.createElement(
          _styledComponents3.Tooltip,
          {
            id: panel.id + '-nav',
            effect: 'solid',
            delayShow: 500,
            place: 'bottom'
          },
          _react2.default.createElement(
            'span',
            null,
            panel.label || panel.id
          )
        )
      );
    })
  );
};

PanelToggle.propTypes = propTypes;

exports.default = PanelToggle;
