// kepler.gl/dist/components/side-panel/panel-header-action.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  margin-left: ', 'px;\n  display: flex;\n  align-items: center;\n  color: ', ';\n\n  :hover {\n    cursor: pointer;\n    color: ', ';\n  }\n\n  &.disabled {\n    pointer-events: none;\n    opacity: 0.3;\n  }\n'], ['\n  margin-left: ', 'px;\n  display: flex;\n  align-items: center;\n  color: ', ';\n\n  :hover {\n    cursor: pointer;\n    color: ', ';\n  }\n\n  &.disabled {\n    pointer-events: none;\n    opacity: 0.3;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames2 = require('classnames');

var _classnames3 = _interopRequireDefault(_classnames2);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _styledComponents3 = require('../common/styled-components');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var HeaderActionWrapper = _styledComponents2.default.div(_templateObject, function (props) {
  return props.flush ? 0 : 8;
}, function (props) {
  return props.active ? props.theme.panelHeaderIconActive : props.theme.panelHeaderIcon;
}, function (props) {
  return props.hoverColor ? props.theme[props.hoverColor] : props.theme.textColorHl;
});

// Need to use react class to access props.component
var PanelHeaderAction = (_temp = _class = function (_Component) {
  (0, _inherits3.default)(PanelHeaderAction, _Component);

  function PanelHeaderAction() {
    (0, _classCallCheck3.default)(this, PanelHeaderAction);
    return (0, _possibleConstructorReturn3.default)(this, (PanelHeaderAction.__proto__ || Object.getPrototypeOf(PanelHeaderAction)).apply(this, arguments));
  }

  (0, _createClass3.default)(PanelHeaderAction, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          onClick = _props.onClick,
          tooltip = _props.tooltip,
          id = _props.id,
          active = _props.active,
          flush = _props.flush,
          hoverColor = _props.hoverColor,
          tooltipType = _props.tooltipType,
          disabled = _props.disabled,
          className = _props.className;

      return _react2.default.createElement(
        HeaderActionWrapper,
        {
          className: (0, _classnames3.default)('panel--header__action', (0, _defineProperty3.default)({ disabled: disabled }, className, className)),
          active: active,
          hoverColor: hoverColor,
          flush: flush
        },
        _react2.default.createElement(this.props.IconComponent, {
          'data-tip': true,
          'data-for': tooltip + '_' + id,
          height: '18px',
          onClick: onClick
        }),
        tooltip ? _react2.default.createElement(
          _styledComponents3.Tooltip,
          {
            id: tooltip + '_' + id,
            effect: 'solid',
            delayShow: 500,
            type: tooltipType
          },
          _react2.default.createElement(
            'span',
            null,
            tooltip
          )
        ) : null
      );
    }
  }]);
  return PanelHeaderAction;
}(_react.Component), _class.propTypes = {
  id: _propTypes2.default.string,
  flush: _propTypes2.default.bool,
  tooltip: _propTypes2.default.string,
  onClick: _propTypes2.default.func,
  active: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  hoverColor: _propTypes2.default.string,
  className: _propTypes2.default.string,
  tooltipType: _propTypes2.default.string
}, _class.defaultProps = {
  onClick: function onClick() {},
  hoverColor: null,
  active: false
}, _temp);
exports.default = PanelHeaderAction;
;
