// kepler.gl/dist/components/side-panel/side-nav.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styledComponents = require('../common/styled-components');

var _defaultSettings = require('../../constants/default-settings');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var SideNav = function SideNav(itemProps) {
  return _react2.default.createElement(
    'div',
    { className: (0, _classnames2.default)('side-navi', { collapsed: itemProps.isCollapsed }) },
    _react2.default.createElement(
      'div',
      null,
      _react2.default.createElement(
        'ul',
        null,
        _defaultSettings.PANELS.map(function (item) {
          return _react2.default.createElement(NavItem, (0, _extends3.default)({}, itemProps, { key: item.id, item: item }));
        })
      )
    ),
    _react2.default.createElement(
      'div',
      null,
      _react2.default.createElement(
        'ul',
        null,
        _defaultSettings.PANELS_FOOTER.map(function (item) {
          return _react2.default.createElement(NavItem, (0, _extends3.default)({}, itemProps, { key: item.id, item: item }));
        })
      )
    )
  );
};

var NavItem = function NavItem(_ref) {
  var activePanel = _ref.activePanel,
      item = _ref.item,
      isCollapsed = _ref.isCollapsed,
      togglePanel = _ref.togglePanel;
  return _react2.default.createElement(
    'li',
    {
      className: (0, _classnames2.default)('side-navi__item', { active: activePanel === item.id }),
      onClick: function onClick() {
        return togglePanel(item.id);
      }
    },
    _react2.default.createElement(
      'a',
      { className: 'hover align--middle', 'data-tip': true, 'data-for': item.id + '-nav' },
      _react2.default.createElement('span', { className: 'icon icon_' + item.icon }),
      isCollapsed ? _react2.default.createElement(
        _styledComponents.Tooltip,
        {
          id: item.id + '-nav',
          effect: 'solid',
          delayShow: 500,
          place: 'right'
        },
        _react2.default.createElement(
          'span',
          null,
          item.label
        )
      ) : _react2.default.createElement(
        'span',
        { className: 'side-navi__item__title' },
        item.label
      )
    )
  );
};

exports.default = SideNav;
