// kepler.gl/dist/components/common/color-legend.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp2;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  ', ';\n\n  max-height: 150px;\n  overflow-y: auto;\n\n  svg {\n    text {\n      font-size: 9px;\n      fill: ', ';\n    }\n  }\n'], ['\n  ', ';\n\n  max-height: 150px;\n  overflow-y: auto;\n\n  svg {\n    text {\n      font-size: 9px;\n      fill: ', ';\n    }\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _reselect = require('reselect');

var _d3Format = require('d3-format');

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _defaultSettings = require('../../constants/default-settings');

var _filterUtils = require('../../utils/filter-utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ROW_H = 10;
var GAP = 4;
var RECT_W = 20;

var StyledLegend = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.sidePanelScrollBar;
}, function (props) {
  return props.theme.textColor;
});

var defaultFormat = function defaultFormat(d) {
  return d;
};

var getTimeLabelFormat = function getTimeLabelFormat(domain) {
  var formatter = (0, _filterUtils.getTimeWidgetHintFormatter)(domain);
  return function (val) {
    return _moment2.default.utc(val).format(formatter);
  };
};

var getNumericLabelFormat = function getNumericLabelFormat(domain) {
  var diff = domain[1] - domain[0];

  if (diff < 10) {
    return (0, _d3Format.format)('.2f');
  }

  return (0, _d3Format.format)('.1f');
};

var getQuantLabelFormat = function getQuantLabelFormat(domain, fieldType) {
  // quant scale can only be assigned to linear Fields: real, timestamp, integer
  return fieldType === _defaultSettings.ALL_FIELD_TYPES.timestamp ? getTimeLabelFormat(domain) : !fieldType ? defaultFormat : getNumericLabelFormat(domain);
};

var getOrdinalLegends = function getOrdinalLegends(scale) {
  var domain = scale.domain();
  return {
    data: domain.map(scale),
    labels: domain
  };
};

var getQuantLegends = function getQuantLegends(scale, labelFormat) {
  var labels = scale.range().map(function (d) {
    var invert = scale.invertExtent(d);
    return labelFormat(invert[0]) + ' to ' + labelFormat(invert[1]);
  });

  return {
    data: scale.range(),
    labels: labels
  };
};

var ColorLegend = (_temp2 = _class = function (_Component) {
  (0, _inherits3.default)(ColorLegend, _Component);

  function ColorLegend() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ColorLegend);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ColorLegend.__proto__ || Object.getPrototypeOf(ColorLegend)).call.apply(_ref, [this].concat(args))), _this), _this.domainSelector = function (props) {
      return props.domain;
    }, _this.rangeSelector = function (props) {
      return props.range;
    }, _this.labelFormatSelector = function (props) {
      return props.labelFormat;
    }, _this.scaleTypeSelector = function (props) {
      return props.scaleType;
    }, _this.fieldTypeSelector = function (props) {
      return props.fieldType;
    }, _this.legendsSelector = (0, _reselect.createSelector)(_this.domainSelector, _this.rangeSelector, _this.scaleTypeSelector, _this.labelFormatSelector, _this.fieldTypeSelector, function (domain, range, scaleType, labelFormat, fieldType) {
      var scaleFunction = _defaultSettings.SCALE_FUNC[scaleType];
      // color scale can only be quantize, quantile or ordinal
      var scale = scaleFunction().domain(domain).range(range);

      if (scaleType === _defaultSettings.SCALE_TYPES.ordinal) {
        return getOrdinalLegends(scale);
      }

      var formatLabel = labelFormat || getQuantLabelFormat(scale.domain(), fieldType);

      return getQuantLegends(scale, formatLabel);
    }), _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(ColorLegend, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          width = _props.width,
          scaleType = _props.scaleType,
          domain = _props.domain,
          range = _props.range,
          _props$displayLabel = _props.displayLabel,
          displayLabel = _props$displayLabel === undefined ? true : _props$displayLabel;


      if (!domain || !range || !scaleType) {
        return null;
      }

      var legends = this.legendsSelector(this.props);
      var height = legends.data.length * (ROW_H + GAP);

      return _react2.default.createElement(
        StyledLegend,
        null,
        _react2.default.createElement(
          'svg',
          { width: width - 24, height: height },
          legends.data.map(function (color, idx) {
            return _react2.default.createElement(LegendRow, {
              key: idx,
              label: legends.labels[idx],
              displayLabel: displayLabel,
              color: color,
              idx: idx
            });
          })
        )
      );
    }
  }]);
  return ColorLegend;
}(_react.Component), _class.propTypes = {
  width: _propTypes2.default.number.isRequired,
  scaleType: _propTypes2.default.string,
  domain: _propTypes2.default.oneOfType([_propTypes2.default.array, _propTypes2.default.object]),
  fieldType: _propTypes2.default.string,
  range: _propTypes2.default.arrayOf(_propTypes2.default.string),
  labelFormat: _propTypes2.default.func
}, _temp2);
exports.default = ColorLegend;


var LegendRow = function LegendRow(_ref2) {
  var _ref2$label = _ref2.label,
      label = _ref2$label === undefined ? '' : _ref2$label,
      displayLabel = _ref2.displayLabel,
      color = _ref2.color,
      idx = _ref2.idx;
  return _react2.default.createElement(
    'g',
    { transform: 'translate(0, ' + idx * (ROW_H + GAP) + ')' },
    _react2.default.createElement('rect', { width: RECT_W, height: ROW_H, style: { fill: color } }),
    _react2.default.createElement(
      'text',
      { x: RECT_W + 8, y: ROW_H - 1 },
      displayLabel ? label.toString() : ''
    )
  );
};
