// kepler.gl/dist/components/common/map-layer-selector.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  padding: 12px;\n\n  .map-layer-selector__item {\n    margin: 12px 0;\n  }\n'], ['\n  padding: 12px;\n\n  .map-layer-selector__item {\n    margin: 12px 0;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _switch = require('./switch');

var _switch2 = _interopRequireDefault(_switch);

var _utils = require('../../utils/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  // Required
  layers: _propTypes2.default.array.isRequired,
  onMapToggleLayer: _propTypes2.default.func.isRequired
};

var MapLayerSelect = _styledComponents2.default.div(_templateObject);

var MapLayerSelector = function MapLayerSelector(_ref) {
  var layers = _ref.layers,
      onMapToggleLayer = _ref.onMapToggleLayer;
  return _react2.default.createElement(
    MapLayerSelect,
    { className: 'map-layer-selector' },
    layers.map(function (layer, index) {
      return _react2.default.createElement(
        'div',
        { key: layer.id, className: 'map-layer-selector__item' },
        _react2.default.createElement(_switch2.default, {
          checked: layer.isVisible,
          id: layer.id + '-toggle-' + (0, _utils.generateHashId)(4),
          label: layer.name,
          onChange: function onChange(e) {
            e.preventDefault();
            onMapToggleLayer(layer.id);
          }
        })
      );
    })
  );
};

MapLayerSelector.propTypes = propTypes;

exports.default = MapLayerSelector;
