// kepler.gl/dist/components/common/modal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalFooter = exports.ModalTitle = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  width: 60%;\n  max-width: 960px;\n  padding: 40px 96px;\n  position: absolute;\n  top: 92px;\n  left: 0;\n  right: 0;\n  margin-left: auto;\n  background-color: #ffffff;\n  border-radius: 4px;\n  transition: ', ';\n  min-width: 600px;\n  overflow: hidden;\n  box-sizing: border-box;\n  margin-right: auto;\n  font-size: 12px;\n  color: ', ';\n  ', ';\n'], ['\n  width: 60%;\n  max-width: 960px;\n  padding: 40px 96px;\n  position: absolute;\n  top: 92px;\n  left: 0;\n  right: 0;\n  margin-left: auto;\n  background-color: #ffffff;\n  border-radius: 4px;\n  transition: ', ';\n  min-width: 600px;\n  overflow: hidden;\n  box-sizing: border-box;\n  margin-right: auto;\n  font-size: 12px;\n  color: ', ';\n  ', ';\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  color: ', ';\n  display: flex;\n  justify-content: flex-end;\n  left: 0;\n  padding: 24px;\n  position: absolute;\n  top: 0;\n  width: 100%;\n  z-index: 10005;\n\n  :hover {\n    cursor: pointer;\n  }\n'], ['\n  color: ', ';\n  display: flex;\n  justify-content: flex-end;\n  left: 0;\n  padding: 24px;\n  position: absolute;\n  top: 0;\n  width: 100%;\n  z-index: 10005;\n\n  :hover {\n    cursor: pointer;\n  }\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  font-size: ', ';\n  color: ', ';\n  margin-bottom: 10px;\n  position: relative;\n  z-index: 10003;\n'], ['\n  font-size: ', ';\n  color: ', ';\n  margin-bottom: 10px;\n  position: relative;\n  z-index: 10003;\n']),
    _templateObject4 = (0, _taggedTemplateLiteral3.default)(['\n  width: 100%;\n  left: 0;\n  bottom: 0;\n  display: flex;\n  flex-direction: column;\n  justify-content: flex-end;\n  padding-top: 36px;\n  z-index: 10001;\n'], ['\n  width: 100%;\n  left: 0;\n  bottom: 0;\n  display: flex;\n  flex-direction: column;\n  justify-content: flex-end;\n  padding-top: 36px;\n  z-index: 10001;\n']),
    _templateObject5 = (0, _taggedTemplateLiteral3.default)(['\n  position: relative;\n  z-index: 10002;\n'], ['\n  position: relative;\n  z-index: 10002;\n']),
    _templateObject6 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  justify-content: flex-end;\n'], ['\n  display: flex;\n  justify-content: flex-end;\n']),
    _templateObject7 = (0, _taggedTemplateLiteral3.default)(['\n  width: 100%;\n  height: 100%;\n  top: 0;\n  left: 0;\n  z-index: 10000;\n  transition: ', ';\n\n  :focus {\n    outline: 0\n  }\n'], ['\n  width: 100%;\n  height: 100%;\n  top: 0;\n  left: 0;\n  z-index: 10000;\n  transition: ', ';\n\n  :focus {\n    outline: 0\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _reactModal = require('react-modal');

var _reactModal2 = _interopRequireDefault(_reactModal);

var _icons = require('./icons');

var _styledComponents3 = require('./styled-components');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ModalContentWrapper = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.transition;
}, function (props) {
  return props.theme.labelColorLT;
}, function (props) {
  return props.cssStyle || '';
});


var CloseButton = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.theme.titleColorLT;
});


var ModalTitle = exports.ModalTitle = _styledComponents2.default.div(_templateObject3, function (props) {
  return props.theme.modalTitleFontSize;
}, function (props) {
  return props.theme.modalTitleColor;
});

var StyledModalFooter = _styledComponents2.default.div(_templateObject4);

var ModalContent = _styledComponents2.default.div(_templateObject5);

var FooterActionWrapper = _styledComponents2.default.div(_templateObject6);

var defaultCancelButton = {
  link: true,
  large: true,
  children: 'Cancel'
};

var defaultConfirmButton = {
  large: true,
  width: '160px',
  children: 'Confirm'
};

var ModalFooter = exports.ModalFooter = function ModalFooter(_ref) {
  var cancel = _ref.cancel,
      confirm = _ref.confirm,
      cancelButton = _ref.cancelButton,
      confirmButton = _ref.confirmButton;

  var cancelButtonProps = (0, _extends3.default)({}, defaultCancelButton, cancelButton);
  var confirmButtonProps = (0, _extends3.default)({}, defaultConfirmButton, confirmButton);
  return _react2.default.createElement(
    StyledModalFooter,
    { className: 'modal--footer' },
    _react2.default.createElement(
      FooterActionWrapper,
      null,
      _react2.default.createElement(
        _styledComponents3.Button,
        (0, _extends3.default)({}, cancelButtonProps, { onClick: cancel }),
        cancelButtonProps.children
      ),
      _react2.default.createElement(
        _styledComponents3.Button,
        (0, _extends3.default)({}, confirmButtonProps, { onClick: confirm }),
        confirmButtonProps.children
      )
    )
  );
};

var ModalDialog = (_temp = _class = function (_Component) {
  (0, _inherits3.default)(ModalDialog, _Component);

  function ModalDialog() {
    (0, _classCallCheck3.default)(this, ModalDialog);
    return (0, _possibleConstructorReturn3.default)(this, (ModalDialog.__proto__ || Object.getPrototypeOf(ModalDialog)).apply(this, arguments));
  }

  (0, _createClass3.default)(ModalDialog, [{
    key: 'render',
    value: function render() {
      var props = this.props;

      return _react2.default.createElement(
        _reactModal2.default,
        (0, _extends3.default)({}, props, {
          ariaHideApp: false,
          style: {
            overlay: {
              backgroundColor: 'rgba(0, 0, 0, 0.5)',
              zIndex: 10000,
              overflowY: 'auto',
              position: 'absolute'
            }
          }
        }),
        _react2.default.createElement(
          ModalContentWrapper,
          {
            className: 'modal--content',
            cssStyle: props.cssStyle,
            footer: props.footer
          },
          props.close ? _react2.default.createElement(
            CloseButton,
            { className: 'modal--close', onClick: props.close },
            _react2.default.createElement(_icons.Delete, { height: '14px' })
          ) : null,
          props.title ? _react2.default.createElement(
            ModalTitle,
            { className: 'modal--title' },
            props.title
          ) : null,
          _react2.default.createElement(
            ModalContent,
            { className: 'content' },
            props.children
          ),
          props.footer ? _react2.default.createElement(ModalFooter, {
            cancel: props.close,
            confirm: props.onConfirm,
            cancelButton: props.cancelButton,
            confirmButton: props.confirmButton
          }) : null
        )
      );
    }
  }]);
  return ModalDialog;
}(_react.Component), _class.propTypes = {
  footer: _propTypes2.default.bool,
  close: _propTypes2.default.func.isRequired,
  onConfirm: _propTypes2.default.func,
  onCancel: _propTypes2.default.func,
  confirmButton: _propTypes2.default.object,
  confirmButtonLabel: _propTypes2.default.string,
  cancelButton: _propTypes2.default.object,
  cancelButtonLabel: _propTypes2.default.string,
  cssStyle: _propTypes2.default.arrayOf(_propTypes2.default.any)
}, _class.defaultProps = {
  footer: false,
  close: function close() {},
  onConfirm: function onConfirm() {},
  onCancel: function onCancel() {},
  cancelButton: defaultCancelButton,
  confirmButton: defaultConfirmButton,
  cssStyle: []
}, _temp);


var StyledModal = (0, _styledComponents2.default)(ModalDialog)(_templateObject7, function (props) {
  return props.theme.transition;
});

exports.default = StyledModal;
