// kepler.gl/dist/components/common/range-brush.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  .selection {\n    stroke: none;\n    fill: ', ';\n    opacity: 1;\n  }\n'], ['\n  .selection {\n    stroke: none;\n    fill: ', ';\n    opacity: 1;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _d3Selection = require('d3-selection');

var _d3Brush = require('d3-brush');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StyledG = _styledComponents2.default.g(_templateObject, function (props) {
  return props.theme.rangeBrushBgd;
});

var RangeBrush = (_temp = _class = function (_Component) {
  (0, _inherits3.default)(RangeBrush, _Component);

  function RangeBrush() {
    (0, _classCallCheck3.default)(this, RangeBrush);
    return (0, _possibleConstructorReturn3.default)(this, (RangeBrush.__proto__ || Object.getPrototypeOf(RangeBrush)).apply(this, arguments));
  }

  (0, _createClass3.default)(RangeBrush, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      var _props = this.props,
          _props$range = (0, _slicedToArray3.default)(_props.range, 2),
          min = _props$range[0],
          max = _props$range[1],
          _props$value = (0, _slicedToArray3.default)(_props.value, 2),
          val0 = _props$value[0],
          val1 = _props$value[1];
      // We want the React app to respond to brush state and vice-versa
      // but d3-brush fires the same events for both user-initiated brushing
      // and programmatic brushing (brush.move). We need these flags to
      // distinguish between the uses.
      //
      // We don't use state because that would trigger another `componentDidUpate`


      this.brushing = false;
      this.moving = false;

      this.root = (0, _d3Selection.select)(this.rootContainer);
      this.brush = (0, _d3Brush.brushX)().on('start', function () {
        _this2.brushing = true;
      }).on('brush', function () {
        if (_this2.moving) {
          return;
        }

        _d3Selection.event.selection === null ? _this2._reset() : _this2._brush(_d3Selection.event.selection);
      }).on('end', function () {
        if (!_this2.moving && _d3Selection.event.selection === null) {
          _this2._reset();
        }

        _this2.brushing = false;
        _this2.moving = false;
      });

      this.root.call(this.brush);

      if (val0 === min && val1 === max) {
        this._reset();
      }
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate(prevProps) {
      var _props2 = this.props,
          _props2$range = (0, _slicedToArray3.default)(_props2.range, 2),
          min = _props2$range[0],
          max = _props2$range[1],
          _props2$value = (0, _slicedToArray3.default)(_props2.value, 2),
          val0 = _props2$value[0],
          val1 = _props2$value[1],
          width = _props2.width;

      var _prevProps$value = (0, _slicedToArray3.default)(prevProps.value, 2),
          prevVal0 = _prevProps$value[0],
          prevVal1 = _prevProps$value[1];

      if (prevProps.width !== width) {
        this.root.call(this.brush);
        this._move(val0, val1);
      }

      if (!this.brushing && !this.moving) {
        if (val0 === min && val1 === max) {
          this.moving = true;
          this.brush.move(this.root, null);
        }

        if (prevVal0 !== val0 || prevVal1 !== val1) {
          this.moving = true;
          this._move(val0, val1);
        }
      }
    }
  }, {
    key: '_reset',
    value: function _reset() {
      var _props$range2 = (0, _slicedToArray3.default)(this.props.range, 2),
          minValue = _props$range2[0],
          maxValue = _props$range2[1];

      this.props.onBrush(minValue, maxValue);
    }
  }, {
    key: '_move',
    value: function _move(val0, val1) {
      var _props3 = this.props,
          _props3$domain = (0, _slicedToArray3.default)(_props3.domain, 2),
          min = _props3$domain[0],
          max = _props3$domain[1],
          width = _props3.width;

      var scale = function scale(x) {
        return (x - min) * width / (max - min);
      };
      this.brush.move(this.root, [scale(val0), scale(val1)]);
    }
  }, {
    key: '_brush',
    value: function _brush(_ref) {
      var _ref2 = (0, _slicedToArray3.default)(_ref, 2),
          sel0 = _ref2[0],
          sel1 = _ref2[1];

      var _props4 = this.props,
          _props4$domain = (0, _slicedToArray3.default)(_props4.domain, 2),
          min = _props4$domain[0],
          max = _props4$domain[1],
          onBrush = _props4.onBrush,
          width = _props4.width;

      var invert = function invert(x) {
        return x * (max - min) / width + min;
      };
      onBrush(invert(sel0), invert(sel1));
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      return _react2.default.createElement(StyledG, { className: 'kg-range-slider__brush',
        innerRef: function innerRef(comp) {
          _this3.rootContainer = comp;
        } });
    }
  }]);
  return RangeBrush;
}(_react.Component), _class.propTypes = {
  domain: _propTypes2.default.arrayOf(_propTypes2.default.number).isRequired,
  onBrush: _propTypes2.default.func.isRequired,
  range: _propTypes2.default.arrayOf(_propTypes2.default.number).isRequired,
  value: _propTypes2.default.arrayOf(_propTypes2.default.number).isRequired,
  width: _propTypes2.default.number.isRequired
}, _temp);
exports.default = RangeBrush;
;
