// kepler.gl/dist/components/common/range-plot.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp2; // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  .rv-xy-plot__inner path {\n    fill: none;\n    stroke-width: 1.5;\n  }\n'], ['\n  .rv-xy-plot__inner path {\n    fill: none;\n    stroke-width: 1.5;\n  }\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  background-color: #d3d8e0;\n  border-radius: 2px;\n  color: ', ';\n  font-size: 9px;\n  margin: 4px;\n  padding: 3px 6px;\n  pointer-events: none;\n  user-select: none;\n'], ['\n  background-color: #d3d8e0;\n  border-radius: 2px;\n  color: ', ';\n  font-size: 9px;\n  margin: 4px;\n  padding: 3px 6px;\n  pointer-events: none;\n  user-select: none;\n']);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _d3Scale = require('d3-scale');

var _moment = require('moment');

var _moment2 = _interopRequireDefault(_moment);

var _d3Array = require('d3-array');

var _reselect = require('reselect');

var _reactVis = require('react-vis');

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _rangeBrush = require('./range-brush');

var _rangeBrush2 = _interopRequireDefault(_rangeBrush);

var _filterUtils = require('../../utils/filter-utils');

var _base = require('../../styles/base');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var chartMargin = { top: 18, bottom: 0, left: 0, right: 0 };
var chartH = 52;
var containerH = 78;
var histogramStyle = {
  highlightW: 0.7,
  unHighlightedW: 0.4,
  highlightedColor: _base.theme.activeColor,
  unHighlightedColor: _base.theme.sliderBarColor
};

var RangePlot = (_temp2 = _class = function (_Component) {
  (0, _inherits3.default)(RangePlot, _Component);

  function RangePlot() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, RangePlot);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = RangePlot.__proto__ || Object.getPrototypeOf(RangePlot)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      hoveredDP: null
    }, _this.domainSelector = function (props) {
      return props.lineChart && props.lineChart.xDomain;
    }, _this.hintFormatter = (0, _reselect.createSelector)(_this.domainSelector, function (domain) {
      return (0, _filterUtils.getTimeWidgetHintFormatter)(domain);
    }), _this.onMouseMove = function (hoveredDP) {
      _this.setState({ hoveredDP: hoveredDP });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(RangePlot, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          onBrush = _props.onBrush,
          range = _props.range,
          value = _props.value,
          width = _props.width,
          plotType = _props.plotType,
          lineChart = _props.lineChart,
          histogram = _props.histogram;

      var domain = [histogram[0].x0, histogram[histogram.length - 1].x1];

      var brushComponent = _react2.default.createElement(_rangeBrush2.default, {
        domain: domain,
        onBrush: onBrush,
        range: range,
        value: value,
        width: width
      });

      return _react2.default.createElement(
        'div',
        {
          style: {
            height: containerH + 'px',
            position: 'relative'
          }
        },
        plotType === 'lineChart' ? _react2.default.createElement(LineChart, {
          hoveredDP: this.state.hoveredDP,
          width: width,
          height: containerH,
          margin: chartMargin,
          children: brushComponent,
          onMouseMove: this.onMouseMove,
          yDomain: lineChart.yDomain,
          hintFormat: this.hintFormatter(this.props),
          data: lineChart.series
        }) : _react2.default.createElement(Histogram, {
          width: width,
          height: chartH,
          value: value,
          margin: chartMargin,
          histogram: histogram,
          brushComponent: brushComponent
        })
      );
    }
  }]);
  return RangePlot;
}(_react.Component), _class.propTypes = {
  value: _propTypes2.default.arrayOf(_propTypes2.default.number).isRequired,
  histogram: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    x0: _propTypes2.default.number,
    x1: _propTypes2.default.number
  })),
  lineChart: _propTypes2.default.object,
  plotType: _propTypes2.default.string,
  isEnlarged: _propTypes2.default.bool,
  onBlur: _propTypes2.default.func,
  width: _propTypes2.default.number.isRequired
}, _temp2);
exports.default = RangePlot;


var Histogram = function Histogram(_ref2) {
  var width = _ref2.width,
      height = _ref2.height,
      margin = _ref2.margin,
      histogram = _ref2.histogram,
      value = _ref2.value,
      brushComponent = _ref2.brushComponent;

  var domain = [histogram[0].x0, histogram[histogram.length - 1].x1];
  var barWidth = width / histogram.length;

  var x = (0, _d3Scale.scaleLinear)().domain(domain).range([0, width]);

  var y = (0, _d3Scale.scaleLinear)().domain([0, (0, _d3Array.max)(histogram, function (d) {
    return d.count;
  })]).range([0, height]);

  return _react2.default.createElement(
    'svg',
    { width: width, height: height, style: { marginTop: margin.top + 'px' } },
    _react2.default.createElement(
      'g',
      { className: 'histogram-bars' },
      histogram.map(function (bar) {
        var inRange = bar.x0 >= value[0] && bar.x1 <= value[1];
        var fill = inRange ? histogramStyle.highlightedColor : histogramStyle.unHighlightedColor;
        var wRatio = inRange ? histogramStyle.highlightW : histogramStyle.unHighlightedW;

        return _react2.default.createElement('rect', {
          key: bar.x0,
          fill: fill,
          height: y(bar.count),
          width: barWidth * wRatio,
          x: x(bar.x0) + barWidth * (1 - wRatio) / 2,
          rx: 1,
          ry: 1,
          y: height - y(bar.count)
        });
      })
    ),
    brushComponent
  );
};

var LineChartWrapper = _styledComponents2.default.div(_templateObject);

var LineChart = function LineChart(_ref3) {
  var width = _ref3.width,
      height = _ref3.height,
      yDomain = _ref3.yDomain,
      hintFormat = _ref3.hintFormat,
      hoveredDP = _ref3.hoveredDP,
      margin = _ref3.margin,
      color = _ref3.color,
      data = _ref3.data,
      onMouseMove = _ref3.onMouseMove,
      children = _ref3.children;

  var brushData = [{ x: data[0].x, y: yDomain[1], customComponent: function customComponent() {
      return children;
    } }];

  return _react2.default.createElement(
    LineChartWrapper,
    null,
    _react2.default.createElement(
      _reactVis.XYPlot,
      { width: width, height: height, margin: (0, _extends3.default)({}, margin, { bottom: 12 }) },
      _react2.default.createElement(_reactVis.LineSeries, {
        strokeWidth: 2,
        color: color,
        data: data,
        onNearestX: onMouseMove
      }),
      _react2.default.createElement(_reactVis.MarkSeries, {
        data: hoveredDP ? [hoveredDP] : [],
        color: color,
        size: 3
      }),
      _react2.default.createElement(_reactVis.CustomSVGSeries, { data: brushData }),
      hoveredDP ? _react2.default.createElement(
        _reactVis.Hint,
        { value: hoveredDP },
        _react2.default.createElement(HintContent, (0, _extends3.default)({}, hoveredDP, {
          format: function format(val) {
            return _moment2.default.utc(val).format(hintFormat);
          }
        }))
      ) : null
    )
  );
};

var StyledHint = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.theme.textColorLT;
});
var HintContent = function HintContent(_ref4) {
  var x = _ref4.x,
      y = _ref4.y,
      format = _ref4.format;
  return _react2.default.createElement(
    StyledHint,
    null,
    _react2.default.createElement(
      'div',
      { className: 'hint--x' },
      format(x)
    ),
    _react2.default.createElement(
      'div',
      { className: 'row' },
      y
    )
  );
};
