// kepler.gl/dist/deckgl-layers/layer-utils/cluster-utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getGeoJSON = getGeoJSON;
exports.clustersAtZoom = clustersAtZoom;
exports.clearClustererCache = clearClustererCache;

var _supercluster = require('supercluster');

var _supercluster2 = _interopRequireDefault(_supercluster);

var _memoize = require('lodash/memoize');

var _memoize2 = _interopRequireDefault(_memoize);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

function getGeoJSON(data, getPosition) {
  return data.map(function (d) {
    return {
      type: 'Point',
      properties: {
        data: d,
        points: [d],
        point_count: 1,
        point_count_abbreviated: '1'
      },
      geometry: {
        coordinates: getPosition(d)
      }
    };
  }).filter(function (d) {
    return d.geometry.coordinates.every(Number.isFinite);
  });
}

var clusterResolver = function clusterResolver(_ref) {
  var clusterRadius = _ref.clusterRadius;
  return '' + clusterRadius;
};

var getClusterer = (0, _memoize2.default)(function (_ref2) {
  var clusterRadius = _ref2.clusterRadius,
      geoJSON = _ref2.geoJSON;

  return (0, _supercluster2.default)({
    maxZoom: 20,
    radius: clusterRadius,
    initial: function initial() {
      return { points: [] };
    },
    map: function map(props) {
      return props.data;
    },
    reduce: function reduce(accumulated, props) {
      if (props.points) {
        // avoid using spread to prevent max call stack exceeded error
        props.points.forEach(function (p) {
          accumulated.points.push(p);
        });
      } else {
        accumulated.points.push(props);
      }
    }
  }).load(geoJSON);
}, clusterResolver);

function clustersAtZoom(_ref3) {
  var bbox = _ref3.bbox,
      clusterRadius = _ref3.clusterRadius,
      geoJSON = _ref3.geoJSON,
      zoom = _ref3.zoom;

  var clusterer = getClusterer({ clusterRadius: clusterRadius, geoJSON: geoJSON });

  return clusterer.getClusters(bbox, zoom);
}

function clearClustererCache() {
  getClusterer.cache.clear();
}
