// @deck.gl/core/dist/es5/experimental/utils/gpu-grid-aggregation/grid-aggregation-utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pointToDensityGridData = pointToDensityGridData;
exports._alignToCell = _alignToCell;

var _assert = _interopRequireDefault(require("assert"));

var _math = require("math.gl");

var _luma = require("luma.gl");

var _constants = require("../../../lib/constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var fp64LowPart = _luma.fp64.fp64LowPart;
var R_EARTH = 6378000; // Takes data and aggregation params and returns aggregated data.

function pointToDensityGridData(_ref) {
  var data = _ref.data,
      getPosition = _ref.getPosition,
      cellSizeMeters = _ref.cellSizeMeters,
      gpuGridAggregator = _ref.gpuGridAggregator,
      gpuAggregation = _ref.gpuAggregation,
      _ref$fp = _ref.fp64,
      fp64 = _ref$fp === void 0 ? false : _ref$fp,
      _ref$coordinateSystem = _ref.coordinateSystem,
      coordinateSystem = _ref$coordinateSystem === void 0 ? _constants.COORDINATE_SYSTEM.LNGLAT : _ref$coordinateSystem,
      _ref$viewport = _ref.viewport,
      viewport = _ref$viewport === void 0 ? null : _ref$viewport;

  var gridData = _parseGridData(data, getPosition);

  var cellSize = [cellSizeMeters, cellSizeMeters];
  var worldOrigin = [0, 0];
  (0, _assert.default)(coordinateSystem === _constants.COORDINATE_SYSTEM.LNGLAT || coordinateSystem === _constants.COORDINATE_SYSTEM.IDENTITY);

  if (coordinateSystem === _constants.COORDINATE_SYSTEM.LNGLAT) {
    // TODO: also for COORDINATE_SYSTEM.LNGLAT_EXPERIMENTAL ?
    var gridOffset = _getGridOffset(gridData, cellSizeMeters);

    cellSize = [gridOffset.xOffset, gridOffset.yOffset];
    worldOrigin = [-180, -90]; // Origin used to define grid cell boundaries
  } else {
    // Other co-ordiante sytems are not verified yet.
    (0, _assert.default)(coordinateSystem === _constants.COORDINATE_SYSTEM.IDENTITY);
    var width = viewport.width,
        height = viewport.height;
    worldOrigin = [-width / 2, -height / 2]; // Origin used to define grid cell boundaries
  }

  var opts = _getGPUAggregationParams({
    gridData: gridData,
    cellSize: cellSize,
    worldOrigin: worldOrigin
  });

  var aggregatedData = gpuGridAggregator.run({
    positions: gridData.positions,
    positions64xyLow: gridData.positions64xyLow,
    weights: gridData.weights,
    cellSize: cellSize,
    width: opts.width,
    height: opts.height,
    gridTransformMatrix: opts.gridTransformMatrix,
    useGPU: gpuAggregation,
    fp64: fp64
  });
  return {
    countsBuffer: aggregatedData.countsBuffer,
    maxCountBuffer: aggregatedData.maxCountBuffer,
    countsData: aggregatedData.countsData,
    maxCountData: aggregatedData.maxCountData,
    gridSize: opts.gridSize,
    gridOrigin: opts.gridOrigin,
    cellSize: cellSize
  };
} // Parse input data to build positions, wights and bounding box.


function _parseGridData(data, getPosition) {
  var getWeight = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
  (0, _assert.default)(data && getPosition);
  var positions = [];
  var positions64xyLow = [];
  var weights = [];
  var yMin = Infinity;
  var yMax = -Infinity;
  var xMin = Infinity;
  var xMax = -Infinity;
  var y;
  var x;

  for (var p = 0; p < data.length; p++) {
    var position = getPosition(data[p]);
    x = position[0];
    y = position[1];
    positions.push(x, y);
    positions64xyLow.push(fp64LowPart(x), fp64LowPart(y));
    var weight = getWeight ? getWeight(data[p]) : 1.0;
    weights.push(weight);

    if (Number.isFinite(y) && Number.isFinite(x)) {
      yMin = y < yMin ? y : yMin;
      yMax = y > yMax ? y : yMax;
      xMin = x < xMin ? x : xMin;
      xMax = x > xMax ? x : xMax;
    }
  }

  return {
    positions: positions,
    positions64xyLow: positions64xyLow,
    weights: weights,
    yMin: yMin,
    yMax: yMax,
    xMin: xMin,
    xMax: xMax
  };
}
/**
 * Based on geometric center of sample points, calculate cellSize in lng/lat (degree) space
 * @param {object} gridData - contains bounding box of data
 * @param {number} cellSize - grid cell size in meters
 * @returns {yOffset, xOffset} - cellSize size lng/lat (degree) space.
 */


function _getGridOffset(gridData, cellSize) {
  var yMin = gridData.yMin,
      yMax = gridData.yMax;
  var latMin = yMin;
  var latMax = yMax;
  var centerLat = (latMin + latMax) / 2;
  return _calculateGridLatLonOffset(cellSize, centerLat);
}
/**
 * calculate grid layer cell size in lat lon based on world unit size
 * and current latitude
 * @param {number} cellSize
 * @param {number} latitude
 * @returns {object} - lat delta and lon delta
 */


function _calculateGridLatLonOffset(cellSize, latitude) {
  var yOffset = _calculateLatOffset(cellSize);

  var xOffset = _calculateLonOffset(latitude, cellSize);

  return {
    yOffset: yOffset,
    xOffset: xOffset
  };
}
/**
 * with a given x-km change, calculate the increment of latitude
 * based on stackoverflow http://stackoverflow.com/questions/7477003
 * @param {number} dy - change in km
 * @return {number} - increment in latitude
 */


function _calculateLatOffset(dy) {
  return dy / R_EARTH * (180 / Math.PI);
}
/**
 * with a given x-km change, and current latitude
 * calculate the increment of longitude
 * based on stackoverflow http://stackoverflow.com/questions/7477003
 * @param {number} lat - latitude of current location (based on city)
 * @param {number} dx - change in km
 * @return {number} - increment in longitude
 */


function _calculateLonOffset(lat, dx) {
  return dx / R_EARTH * (180 / Math.PI) / Math.cos(lat * Math.PI / 180);
} // Aligns `inValue` to given `cellSize`


function _alignToCell(inValue, cellSize) {
  var sign = inValue < 0 ? -1 : 1;
  var value = sign < 0 ? Math.abs(inValue) + cellSize : Math.abs(inValue);
  value = Math.floor(value / cellSize) * cellSize;
  return value * sign;
} // Calculate grid parameters


function _getGPUAggregationParams(_ref2) {
  var gridData = _ref2.gridData,
      cellSize = _ref2.cellSize,
      worldOrigin = _ref2.worldOrigin;
  var yMin = gridData.yMin,
      yMax = gridData.yMax,
      xMin = gridData.xMin,
      xMax = gridData.xMax; // NOTE: this alignment will match grid cell boundaries with existing CPU implementation
  // this gurantees identical aggregation results when switching between CPU and GPU aggregation.
  // Also gurantees same cell boundaries, when overlapping between two different layers (like ScreenGrid and Contour)
  // We first move worldOrigin to [0, 0], align the lower bounding box , then move worldOrigin to its original value.

  var originX = _alignToCell(xMin - worldOrigin[0], cellSize[0]) + worldOrigin[0];
  var originY = _alignToCell(yMin - worldOrigin[1], cellSize[1]) + worldOrigin[1]; // Setup transformation matrix so that every point is in +ve range

  var gridTransformMatrix = new _math.Matrix4().translate([-1 * originX, -1 * originY, 0]); // const cellSize = [gridOffset.xOffset, gridOffset.yOffset];

  var gridOrigin = [originX, originY];
  var width = xMax - xMin + cellSize[0];
  var height = yMax - yMin + cellSize[1];
  var gridSize = [Math.ceil(width / cellSize[0]), Math.ceil(height / cellSize[1])];
  return {
    gridOrigin: gridOrigin,
    gridSize: gridSize,
    width: width,
    height: height,
    gridTransformMatrix: gridTransformMatrix
  };
}
