// probe.gl/dist/es5/test-utils/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "makeSpy", {
  enumerable: true,
  get: function get() {
    return _makeSpy.makeSpy;
  }
});
Object.defineProperty(exports, "BrowserDriver", {
  enumerable: true,
  get: function get() {
    return _browserDriver.default;
  }
});
Object.defineProperty(exports, "BrowserTestDriver", {
  enumerable: true,
  get: function get() {
    return _browserTestDriver.default;
  }
});
Object.defineProperty(exports, "callExposedFunction", {
  enumerable: true,
  get: function get() {
    return _callExposedFunction.callExposedFunction;
  }
});
exports.experimental = void 0;

var _makeSpy = require("./make-spy");

var _browserDriver = _interopRequireDefault(require("./browser-automation/browser-driver"));

var _browserTestDriver = _interopRequireDefault(require("./browser-automation/browser-test-driver"));

var _callExposedFunction = require("./browser-automation/call-exposed-function");

var _loadImage = require("./image-utils/load-image");

var _imageTools = require("./image-utils/image-tools");

var _diffImages = require("./image-utils/diff-images");

var experimental = {
  loadImage: _loadImage.loadImage,
  createImage: _imageTools.createImage,
  getImageFromContext: _imageTools.getImageFromContext,
  getImagePixelData: _imageTools.getImagePixelData,
  diffImagePixels: _diffImages.diffImagePixels
};
exports.experimental = experimental;
