// kepler.gl/dist/components/common/slider/slider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp2;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  position: relative;\n  margin-bottom: 12px;\n  background-color: ', ';\n  height: ', ';\n'], ['\n  position: relative;\n  margin-bottom: 12px;\n  background-color: ', ';\n  height: ', ';\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  flex-grow: 1;\n  margin-top: ', 'px;\n'], ['\n  flex-grow: 1;\n  margin-top: ', 'px;\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _sliderHandle = require('./slider-handle');

var _sliderHandle2 = _interopRequireDefault(_sliderHandle);

var _sliderBarHandle = require('./slider-bar-handle');

var _sliderBarHandle2 = _interopRequireDefault(_sliderBarHandle);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function noop() {}

var StyledRangeSlider = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.sliderBarBgd;
}, function (props) {
  return props.theme.sliderBarHeight;
});

var SliderWrapper = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.isRanged ? 0 : 10;
});

var Slider = (_temp2 = _class = function (_Component) {
  (0, _inherits3.default)(Slider, _Component);

  function Slider() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, Slider);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = Slider.__proto__ || Object.getPrototypeOf(Slider)).call.apply(_ref, [this].concat(args))), _this), _this.ref = undefined, _this._saveRef = function (ref) {
      _this.ref = ref;
    }, _this.slide0Listener = function (x) {
      var xPercent = x / _this.ref.offsetWidth;
      var maxDelta = _this.props.maxValue - _this.props.minValue;
      var val = xPercent * maxDelta;
      _this.props.onSlider0Change.call(_this, val + _this.props.value0);
    }, _this.slide1Listener = function (x) {
      var xPercent = x / _this.ref.offsetWidth;
      var maxDelta = _this.props.maxValue - _this.props.minValue;
      var val = xPercent * maxDelta;
      _this.props.onSlider1Change(val + _this.props.value1);
    }, _this.sliderBarListener = function (x) {
      var xPercent = x / _this.ref.offsetWidth;
      var maxDelta = _this.props.maxValue - _this.props.minValue;
      var val = xPercent * maxDelta;
      var val0 = val + _this.props.value0;
      var val1 = val + _this.props.value1;
      _this.props.onSliderBarChange(val0, val1);
    }, _this.calcHandleLeft0 = function (w, l, num) {
      return w === 0 ? 'calc(' + l + '% - ' + _this.props.sliderHandleWidth / 2 + 'px)' : 'calc(' + l + '% - ' + _this.props.sliderHandleWidth / 2 + 'px)';
    }, _this.calcHandleLeft1 = function (w, l) {
      return _this.props.isRanged && w === 0 ? l + '%' : 'calc(' + (l + w) + '% - ' + _this.props.sliderHandleWidth / 2 + 'px)';
    }, _this.createSlider = function (width, v0Left) {
      return _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          StyledRangeSlider,
          { className: 'kg-range-slider' },
          _react2.default.createElement(_sliderHandle2.default, {
            className: 'kg-range-slider__handle',
            left: _this.calcHandleLeft0(width, v0Left),
            valueListener: _this.slide0Listener,
            sliderHandleWidth: _this.props.sliderHandleWidth,
            display: _this.props.isRanged
          }),
          _react2.default.createElement(_sliderHandle2.default, {
            className: 'kg-range-slider__handle',
            left: _this.calcHandleLeft1(width, v0Left),
            valueListener: _this.slide1Listener,
            sliderHandleWidth: _this.props.sliderHandleWidth
          }),
          _react2.default.createElement(_sliderBarHandle2.default, {
            width: width,
            v0Left: v0Left,
            enableBarDrag: _this.props.enableBarDrag,
            sliderBarListener: _this.sliderBarListener
          })
        )
      );
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(Slider, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          classSet = _props.classSet,
          isRanged = _props.isRanged,
          maxValue = _props.maxValue,
          minValue = _props.minValue,
          value1 = _props.value1;

      var value0 = !isRanged && minValue > 0 ? minValue : this.props.value0;
      var currValDelta = value1 - value0;
      var maxDelta = maxValue - minValue;
      var width = currValDelta / maxDelta * 100;

      var v0Left = (value0 - minValue) / maxDelta * 100;

      return _react2.default.createElement(
        SliderWrapper,
        {
          className: (0, _classnames2.default)('kg-slider', (0, _extends3.default)({}, classSet)),
          innerRef: this._saveRef,
          isRanged: isRanged
        },
        this.createSlider(width, v0Left)
      );
    }
  }]);
  return Slider;
}(_react.Component), _class.propTypes = {
  title: _propTypes2.default.string,
  isRanged: _propTypes2.default.bool,
  value0: _propTypes2.default.number,
  value1: _propTypes2.default.number,
  minValue: _propTypes2.default.number,
  maxValue: _propTypes2.default.number,
  sliderHandleWidth: _propTypes2.default.number,
  onSlider0Change: _propTypes2.default.func,
  onInput0Change: _propTypes2.default.func,
  onSlider1Change: _propTypes2.default.func,
  onInput1Change: _propTypes2.default.func,
  onSliderBarChange: _propTypes2.default.func,
  step: _propTypes2.default.number,
  enableBarDrag: _propTypes2.default.bool
}, _class.defaultProps = {
  title: '',
  isRanged: true,
  value0: 0,
  value1: 100,
  minValue: 0,
  maxValue: 100,
  step: 1,
  sliderHandleWidth: 12,
  enableBarDrag: false,
  onSlider0Change: noop,
  onInput0Change: noop,
  onSlider1Change: noop,
  onInput1Change: noop,
  onSliderBarChange: noop,
  disabled: false
}, _temp2);
exports.default = Slider;
