// kepler.gl/dist/components/bottom-widget.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.default = BottomWidgetFactory;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _timeWidget = require('./filters/time-widget');

var _timeWidget2 = _interopRequireDefault(_timeWidget);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propsTypes = {
  filters: _propTypes2.default.array,
  datasets: _propTypes2.default.object,
  uiState: _propTypes2.default.object,
  visStateActions: _propTypes2.default.object,
  sidePanelWidth: _propTypes2.default.number,
  containerW: _propTypes2.default.number
}; // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var MaxWidth = 1080;

BottomWidgetFactory.deps = [_timeWidget2.default];

function BottomWidgetFactory(TimeWidget) {
  var BottomWidget = function (_Component) {
    (0, _inherits3.default)(BottomWidget, _Component);

    function BottomWidget() {
      (0, _classCallCheck3.default)(this, BottomWidget);
      return (0, _possibleConstructorReturn3.default)(this, (BottomWidget.__proto__ || Object.getPrototypeOf(BottomWidget)).apply(this, arguments));
    }

    (0, _createClass3.default)(BottomWidget, [{
      key: 'render',
      value: function render() {
        var _props = this.props,
            datasets = _props.datasets,
            filters = _props.filters,
            visStateActions = _props.visStateActions,
            containerW = _props.containerW,
            uiState = _props.uiState,
            sidePanelWidth = _props.sidePanelWidth;
        var activeSidePanel = uiState.activeSidePanel;

        var isOpen = Boolean(activeSidePanel);

        var enlargedFilterIdx = filters.findIndex(function (f) {
          return f.enlarged;
        });
        var isAnyFilterAnimating = filters.some(function (f) {
          return f.isAnimating;
        });
        var enlargedFilterWidth = isOpen ? containerW - sidePanelWidth : containerW;

        if (enlargedFilterIdx < 0) {
          return null;
        }

        return _react2.default.createElement(TimeWidget, {
          fields: datasets[filters[enlargedFilterIdx].dataId].fields,
          setFilterPlot: visStateActions.setFilterPlot,
          setFilter: visStateActions.setFilter,
          toggleAnimation: visStateActions.toggleAnimation,
          updateAnimationSpeed: visStateActions.updateAnimationSpeed,
          enlargeFilter: visStateActions.enlargeFilter,
          width: Math.min(MaxWidth, enlargedFilterWidth),
          isAnyFilterAnimating: isAnyFilterAnimating,
          enlargedIdx: enlargedFilterIdx,
          filter: filters[enlargedFilterIdx]
        });
      }
    }]);
    return BottomWidget;
  }(_react.Component);

  BottomWidget.propsTypes = propsTypes;

  return BottomWidget;
}
