// kepler.gl/dist/components/container.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.appInjector = exports.errorMsg = undefined;

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.ContainerFactory = ContainerFactory;
exports.injectComponents = injectComponents;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _lodash = require('lodash.memoize');

var _lodash2 = _interopRequireDefault(_lodash);

var _window = require('global/window');

var _injector = require('./injector');

var _keplerGl = require('./kepler-gl');

var _keplerGl2 = _interopRequireDefault(_keplerGl);

var _actionWrapper = require('../actions/action-wrapper');

var _identityActions = require('../actions/identity-actions');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var errorMsg = exports.errorMsg = {
  noState: 'kepler.gl state doesnt exist. ' + 'You might forget to mount keplerGlReducer in your root reducer.' + 'If it is not mounted as state.keplerGl by default, you need to provide getState as a prop',

  wrongType: function wrongType(type) {
    return 'injectComponents takes an array of factories replacement pairs as input, ' + (type + ' is provided');
  },

  wrongPairType: 'injectComponents takes an array of factories replacement pairs as input, ' + 'each pair be a array as [originalFactory, replacement]'
};

ContainerFactory.deps = [_keplerGl2.default];

function ContainerFactory(KeplerGl) {
  var _class, _temp;

  var Container = (_temp = _class = function (_Component) {
    (0, _inherits3.default)(Container, _Component);

    function Container(props, ctx) {
      (0, _classCallCheck3.default)(this, Container);

      var _this = (0, _possibleConstructorReturn3.default)(this, (Container.__proto__ || Object.getPrototypeOf(Container)).call(this, props, ctx));

      _this.getSelector = (0, _lodash2.default)(function (id, getState) {
        return function (state) {
          if (!getState(state)) {
            // log error
            _window.console.error(errorMsg.noState);

            return null;
          }
          return getState(state)[id];
        };
      });
      _this.getDispatch = (0, _lodash2.default)(function (id, dispatch) {
        return (0, _actionWrapper.forwardTo)(id, dispatch);
      });
      return _this;
    }
    // default id and address if not provided


    (0, _createClass3.default)(Container, [{
      key: 'componentWillMount',
      value: function componentWillMount() {
        var _props = this.props,
            id = _props.id,
            mint = _props.mint,
            mapboxApiAccessToken = _props.mapboxApiAccessToken;
        // add a new entry to reducer

        this.props.dispatch((0, _identityActions.registerEntry)({ id: id, mint: mint, mapboxApiAccessToken: mapboxApiAccessToken }));
      }
    }, {
      key: 'componentWillReceiveProps',
      value: function componentWillReceiveProps(nextProps) {
        // check if id has changed, if true, copy state over
        if (nextProps.id !== this.props.id) {
          this.props.dispatch((0, _identityActions.renameEntry)(this.props.id, nextProps));
        }
      }
    }, {
      key: 'componentWillUnmount',
      value: function componentWillUnmount() {
        if (this.props.mint !== false) {
          // delete entry in reducer
          this.props.dispatch((0, _identityActions.deleteEntry)(this.props.id));
        }
      }
    }, {
      key: 'render',
      value: function render() {
        var _props2 = this.props,
            id = _props2.id,
            getState = _props2.getState,
            dispatch = _props2.dispatch,
            state = _props2.state;

        var selector = this.getSelector(id, getState);

        if (!selector || !selector(state)) {
          // instance state hasn't been mounted yet
          return _react2.default.createElement('div', null);
        }

        return _react2.default.createElement(KeplerGl, (0, _extends3.default)({}, this.props, {
          id: id,
          selector: selector,
          dispatch: this.getDispatch(id, dispatch)
        }));
      }
    }]);
    return Container;
  }(_react.Component), _class.defaultProps = {
    id: 'map',
    getState: function getState(state) {
      return state.keplerGl;
    },
    mint: true
  }, _temp);


  var mapStateToProps = function mapStateToProps(state, props) {
    return (0, _extends3.default)({ state: state }, props);
  };
  var dispatchToProps = function dispatchToProps(dispatch) {
    return { dispatch: dispatch };
  };
  return (0, _reactRedux.connect)(mapStateToProps, dispatchToProps)(Container);
}

// provide all recipes to injector
var appInjector = exports.appInjector = [ContainerFactory].concat((0, _toConsumableArray3.default)(ContainerFactory.deps), (0, _toConsumableArray3.default)(_keplerGl2.default.deps), (0, _toConsumableArray3.default)(_keplerGl.keplerGlChildDeps)).reduce(function (inj, factory) {
  return inj.provide(factory, factory);
}, (0, _injector.injector)());

// Helper to inject custom components and return kepler.gl container
function injectComponents(recipes) {
  if (!Array.isArray(recipes)) {
    _window.console.error(errorMsg.wrongType(typeof recipes === 'undefined' ? 'undefined' : (0, _typeof3.default)(recipes)));
    return appInjector.get(ContainerFactory);
  }

  return recipes.reduce(function (inj, recipe) {
    if (!Array.isArray(recipes)) {
      _window.console.error(errorMsg.wrongPairType);
      return inj;
    }
    return inj.provide.apply(inj, (0, _toConsumableArray3.default)(recipe));
  }, appInjector).get(ContainerFactory);
}

var InjectedContainer = appInjector.get(ContainerFactory);

exports.default = InjectedContainer;
