// kepler.gl/dist/components/injector.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.errorMsg = undefined;

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends4 = require('babel-runtime/helpers/extends');

var _extends5 = _interopRequireDefault(_extends4);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

exports.injector = injector;
exports.withState = withState;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactRedux = require('react-redux');

var _redux = require('redux');

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _window = require('global/window');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MissingComp = function MissingComp() {
  return _react2.default.createElement('div', null);
}; // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var errorMsg = exports.errorMsg = {
  noDep: function noDep(fac, parent) {
    return fac.name + ' is required as a dependency of ' + parent.name + ', ' + 'but is not provided to injectComponents. It will not be rendered';
  },
  notFunc: '`factory and its replacment should be a function`'
};

function injector() {
  var map = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

  var cache = {}; // map<factory, factory -> ?>
  var get = function get(fac, parent) {
    var factory = map[fac];
    // factory is not injected
    if (!factory) {
      _window.console.error(errorMsg.noDep(fac, parent));
      return MissingComp;
    }

    var instances = cache[factory] || factory.apply(undefined, (0, _toConsumableArray3.default)(factory.deps ? factory.deps.map(function (dep) {
      return get(dep, factory);
    }) : []));

    cache[fac] = instances;
    return instances;
  };

  // if you have two functions that happen to have the exactly same text
  // it will be override: 2018-02-05
  return {
    provide: function provide(factory, replacement) {
      if (typeof factory !== 'function' || typeof replacement !== 'function') {
        _window.console.error(errorMsg.notFunc);
        return injector(map);
      }
      return injector((0, _extends5.default)({}, map, (0, _defineProperty3.default)({}, factory, replacement)));
    },
    get: get
  };
}

var identity = function identity(state) {
  return state;
};
// Helper to add reducer state to custom component
function withState(lenses) {
  var mapStateToProps = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : identity;
  var actions = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

  return function (Component) {
    var WrappedComponent = function WrappedComponent(_ref, _ref2) {
      var selector = _ref2.selector,
          id = _ref2.id;
      var state = _ref.state,
          props = (0, _objectWithoutProperties3.default)(_ref, ['state']);
      return _react2.default.createElement(Component, lenses.reduce(function (totalState, lens) {
        return (0, _extends5.default)({}, totalState, lens(selector(state)));
      }, props));
    };
    WrappedComponent.contextTypes = {
      selector: _propTypes2.default.func,
      id: _propTypes2.default.string
    };
    return (0, _reactRedux.connect)(function (state) {
      return (0, _extends5.default)({}, mapStateToProps(state), { state: state });
    }, function (dispatch) {
      return Object.keys(actions).reduce(function (accu, key) {
        return (0, _extends5.default)({}, accu, (0, _defineProperty3.default)({}, key, (0, _redux.bindActionCreators)(actions[key], dispatch)));
      }, {});
    })(WrappedComponent);
  };
}

// Helpter to add actionCreator to custom component
