// kepler.gl/dist/components/modal-container.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  height: 85%;\n  width: 90%;\n  top: 80px;\n  padding: 32px 0 0 0;\n  max-width: unset;\n'], ['\n  height: 85%;\n  width: 90%;\n  top: 80px;\n  padding: 32px 0 0 0;\n  max-width: unset;\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  width: 40%;\n  padding: 40px 40px 32px 40px;\n'], ['\n  width: 40%;\n  padding: 40px 40px 32px 40px;\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  top: 60px;\n'], ['\n  top: 60px;\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

// modals


exports.default = ModalContainerFactory;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _reactDom = require('react-dom');

var _window = require('global/window');

var _modal = require('./common/modal');

var _modal2 = _interopRequireDefault(_modal);

var _dataProcessor = require('../processors/data-processor');

var _schemas = require('../schemas');

var _schemas2 = _interopRequireDefault(_schemas);

var _exportImageUtils = require('../utils/export-image-utils');

var _deleteDataModal = require('./modals/delete-data-modal');

var _deleteDataModal2 = _interopRequireDefault(_deleteDataModal);

var _dataTableModal = require('./modals/data-table-modal');

var _dataTableModal2 = _interopRequireDefault(_dataTableModal);

var _loadDataModal = require('./modals/load-data-modal');

var _loadDataModal2 = _interopRequireDefault(_loadDataModal);

var _exportImageModal = require('./modals/export-image-modal');

var _exportImageModal2 = _interopRequireDefault(_exportImageModal);

var _exportDataModal = require('./modals/export-data-modal');

var _exportDataModal2 = _interopRequireDefault(_exportDataModal);

var _exportConfigModal = require('./modals/export-config-modal');

var _exportConfigModal2 = _interopRequireDefault(_exportConfigModal);

var _addMapStyleModal = require('./modals/add-map-style-modal');

var _addMapStyleModal2 = _interopRequireDefault(_addMapStyleModal);

var _defaultSettings = require('../constants/default-settings');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var DataTableModalStyle = (0, _styledComponents.css)(_templateObject);

var DeleteDatasetModalStyled = (0, _styledComponents.css)(_templateObject2);

var LoadDataModalStyle = (0, _styledComponents.css)(_templateObject3);

ModalContainerFactory.deps = [_deleteDataModal2.default, _dataTableModal2.default, _loadDataModal2.default, _exportImageModal2.default, _exportDataModal2.default, _exportConfigModal2.default, _addMapStyleModal2.default];

function ModalContainerFactory(DeleteDatasetModal, DataTableModal, LoadDataModal, ExportImageModal, ExportDataModal, ExportConfigModal, AddMapStyleModal) {
  var _class, _temp2;

  var ModalWrapper = (_temp2 = _class = function (_Component) {
    (0, _inherits3.default)(ModalWrapper, _Component);

    function ModalWrapper() {
      var _ref;

      var _temp, _this, _ret;

      (0, _classCallCheck3.default)(this, ModalWrapper);

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ModalWrapper.__proto__ || Object.getPrototypeOf(ModalWrapper)).call.apply(_ref, [this].concat(args))), _this), _this._closeModal = function () {
        _this.props.uiStateActions.toggleModal(null);
      }, _this._deleteDataset = function (key) {
        _this.props.visStateActions.removeDataset(key);
        _this._closeModal();
      }, _this._onAddCustomMapStyle = function () {
        _this.props.mapStyleActions.addCustomMapStyle();
        _this._closeModal();
      }, _this._onFileUpload = function (blob) {
        _this.props.visStateActions.loadFiles(blob);
      }, _this._onExportImage = function () {
        var _this$props$uiState$e = _this.props.uiState.exportImage,
            exporting = _this$props$uiState$e.exporting,
            imageDataUri = _this$props$uiState$e.imageDataUri;

        if (!exporting && imageDataUri) {
          var file = (0, _exportImageUtils.dataURItoBlob)(imageDataUri);
          (0, _exportImageUtils.downloadFile)(file, _defaultSettings.DEFAULT_EXPORT_IMAGE_NAME);
        }
        _this.props.uiStateActions.cleanupExportImage();
        _this._closeModal();
      }, _this._onExportData = function () {
        var _this$props = _this.props,
            visState = _this$props.visState,
            uiState = _this$props.uiState;
        var datasets = visState.datasets;
        var _uiState$exportData = uiState.exportData,
            selectedDataset = _uiState$exportData.selectedDataset,
            dataType = _uiState$exportData.dataType,
            filtered = _uiState$exportData.filtered;
        // get the selected data

        var filename = 'kepler-gl';
        var selectedDatasets = datasets[selectedDataset] ? [datasets[selectedDataset]] : Object.values(datasets);
        if (!selectedDatasets.length) {
          // error: selected dataset not found.
          _this._closeModal();
        }

        selectedDatasets.forEach(function (selectedData) {
          var allData = selectedData.allData,
              data = selectedData.data,
              fields = selectedData.fields,
              label = selectedData.label;

          var exportData = filtered ? data : allData;
          // start to export data according to selected data type
          switch (dataType) {
            case _defaultSettings.EXPORT_DATA_TYPE.CSV:
              {
                var type = 'text/csv';
                var csv = (0, _dataProcessor.formatCsv)(exportData, fields);
                _this._downloadFile(csv, type, filename + '_' + label + '.csv');
                break;
              }
            // TODO: support more file types.
            default:
              break;
          }
        });

        _this._closeModal();
      }, _this._onExportConfig = function () {
        var data = _this.props.uiState.exportData.data;

        // we pass all props because we avoid to create new variables

        var dump = data ? _schemas2.default.save(_this.props) : _schemas2.default.getConfigToSave(_this.props);

        _this._downloadFile(JSON.stringify(dump, null, 2), 'application/json', 'keplergl.json');

        _this._closeModal();
      }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
    }

    (0, _createClass3.default)(ModalWrapper, [{
      key: '_downloadFile',
      value: function _downloadFile(data, type, filename) {
        var fileBlob = new _window.Blob([data], { type: type });
        (0, _exportImageUtils.downloadFile)(fileBlob, filename);
      }
    }, {
      key: 'render',
      value: function render() {
        var _this2 = this;

        var _props = this.props,
            containerW = _props.containerW,
            containerH = _props.containerH,
            mapStyle = _props.mapStyle,
            mapState = _props.mapState,
            uiState = _props.uiState,
            visState = _props.visState,
            rootNode = _props.rootNode,
            visStateActions = _props.visStateActions;
        var currentModal = uiState.currentModal,
            datasetKeyToRemove = uiState.datasetKeyToRemove;
        var datasets = visState.datasets,
            layers = visState.layers,
            editingDataset = visState.editingDataset;


        var template = null;
        var modalProps = {};

        if (currentModal && currentModal.id && currentModal.template) {
          // if currentMdoal template is already provided
          // TODO: need to check whether template is valid
          template = _react2.default.createElement(currentModal.template, null);
          modalProps = currentModal.modalProps;
        } else {
          switch (currentModal) {
            case _defaultSettings.DATA_TABLE_ID:
              template = _react2.default.createElement(DataTableModal, {
                width: containerW * 0.9,
                height: containerH * 0.85,
                datasets: datasets,
                dataId: editingDataset,
                showDatasetTable: visStateActions.showDatasetTable
              });
              modalProps.cssStyle = DataTableModalStyle;
              break;
            case _defaultSettings.DELETE_DATA_ID:
              // validate options
              if (datasetKeyToRemove && datasets && datasets[datasetKeyToRemove]) {
                template = _react2.default.createElement(DeleteDatasetModal, {
                  dataset: datasets[datasetKeyToRemove],
                  layers: layers
                });

                modalProps = {
                  title: 'Delete Dataset',
                  cssStyle: DeleteDatasetModalStyled,
                  footer: true,
                  onConfirm: function onConfirm() {
                    return _this2._deleteDataset(datasetKeyToRemove);
                  },
                  onCancel: this._closeModal,
                  confirmButton: {
                    negative: true,
                    large: true,
                    children: 'Delete'
                  }
                };
              }
              break; // in case we add a new case after this one
            case _defaultSettings.ADD_DATA_ID:
              template = _react2.default.createElement(LoadDataModal, {
                onClose: this._closeModal,
                onFileUpload: this._onFileUpload
              });
              modalProps = {
                title: 'Welcome to VisCrime',
                cssStyle: LoadDataModalStyle,
                footer: false,
                onConfirm: this._closeModal
              };
              break;

            case _defaultSettings.EXPORT_IMAGE_ID:
              var _uiState$exportImage = uiState.exportImage,
                  ratio = _uiState$exportImage.ratio,
                  legend = _uiState$exportImage.legend,
                  resolution = _uiState$exportImage.resolution,
                  exporting = _uiState$exportImage.exporting,
                  imageDataUri = _uiState$exportImage.imageDataUri;

              template = _react2.default.createElement(ExportImageModal, {
                width: containerW,
                height: containerH,
                legend: legend,
                ratio: ratio,
                resolution: resolution,
                exporting: exporting,
                imageDataUri: imageDataUri,
                onChangeRatio: this.props.uiStateActions.setRatio,
                onChangeResolution: this.props.uiStateActions.setResolution,
                onToggleLegend: this.props.uiStateActions.toggleLegend
              });
              modalProps = {
                close: false,
                title: 'Export Image',
                footer: true,
                onCancel: this._closeModal,
                onConfirm: this._onExportImage,
                confirmButton: {
                  large: true,
                  disabled: exporting,
                  children: 'Download'
                }
              };
              break;

            case _defaultSettings.EXPORT_DATA_ID:

              template = _react2.default.createElement(ExportDataModal, (0, _extends3.default)({}, uiState.exportData, {
                datasets: datasets,
                onClose: this._closeModal,
                onChangeExportDataType: this.props.uiStateActions.setExportDataType,
                onChangeExportSelectedDataset: this.props.uiStateActions.setExportSelectedDataset,
                onChangeExportFiltered: this.props.uiStateActions.setExportFiltered
              }));
              modalProps = {
                close: false,
                title: 'Export Data',
                footer: true,
                onCancel: this._closeModal,
                onConfirm: this._onExportData,
                confirmButton: {
                  large: true,
                  children: 'Export'
                }
              };
              break;

            case _defaultSettings.EXPORT_CONFIG_ID:
              var keplerGlConfig = _schemas2.default.getConfigToSave({ mapStyle: mapStyle, visState: visState, mapState: mapState, uiState: uiState });
              template = _react2.default.createElement(ExportConfigModal, {
                config: keplerGlConfig,
                data: uiState.exportData.data,
                onClose: this._closeModal,
                onChangeExportData: this.props.uiStateActions.setExportData
              });
              modalProps = {
                close: false,
                title: 'Export Config',
                footer: true,
                onCancel: this._closeModal,
                onConfirm: this._onExportConfig,
                confirmButton: {
                  large: true,
                  children: 'Export'
                }
              };
              break;

            case _defaultSettings.ADD_MAP_STYLE_ID:
              template = _react2.default.createElement(AddMapStyleModal, {
                mapboxApiAccessToken: this.props.mapboxApiAccessToken,
                mapState: this.props.mapState,
                inputStyle: mapStyle.inputStyle,
                inputMapStyle: this.props.mapStyleActions.inputMapStyle,
                loadCustomMapStyle: this.props.mapStyleActions.loadCustomMapStyle
              });
              modalProps = {
                close: false,
                title: 'Add Custom Mapbox Style',
                footer: true,
                onCancel: this._closeModal,
                onConfirm: this._onAddCustomMapStyle,
                confirmButton: {
                  large: true,
                  disabled: !mapStyle.inputStyle.style,
                  children: 'Add Style'
                }
              };
              break;
            default:
              break;
          }
        }

        return this.props.rootNode ? _react2.default.createElement(
          _modal2.default,
          (0, _extends3.default)({}, modalProps, {
            parentSelector: function parentSelector() {
              return (0, _reactDom.findDOMNode)(rootNode);
            },
            isOpen: Boolean(currentModal),
            close: this._closeModal
          }),
          template
        ) : null;
      }
    }]);
    return ModalWrapper;
  }(_react.Component), _class.propTypes = {
    rootNode: _propTypes2.default.object,
    containerW: _propTypes2.default.number,
    containerH: _propTypes2.default.number,
    mapboxApiAccessToken: _propTypes2.default.string.isRequired,
    mapState: _propTypes2.default.object.isRequired,
    mapStyle: _propTypes2.default.object.isRequired,
    uiState: _propTypes2.default.object.isRequired,
    visState: _propTypes2.default.object.isRequired,
    visStateActions: _propTypes2.default.object.isRequired,
    uiStateActions: _propTypes2.default.object.isRequired,
    mapStyleActions: _propTypes2.default.object.isRequired
  }, _temp2);


  return ModalWrapper;
}
