// kepler.gl/dist/components/plot-container.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  .mapboxgl-ctrl-bottom-left,\n  .mapboxgl-ctrl-bottom-right {\n    display: none;\n  }\n'], ['\n  .mapboxgl-ctrl-bottom-left,\n  .mapboxgl-ctrl-bottom-right {\n    display: none;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

// libraries


exports.default = PlotContainerFactory;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reselect = require('reselect');

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _reactMapGl = require('react-map-gl');

var _lodash = require('lodash.debounce');

var _lodash2 = _interopRequireDefault(_lodash);

var _window = require('global/window');

var _window2 = _interopRequireDefault(_window);

var _mapContainer = require('./map-container');

var _mapContainer2 = _interopRequireDefault(_mapContainer);

var _exportImageUtils = require('../utils/export-image-utils');

var _mapboxGlStyleEditor = require('../utils/map-style-utils/mapbox-gl-style-editor');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  width: _propTypes2.default.number.isRequired,
  height: _propTypes2.default.number.isRequired,
  exportImageSetting: _propTypes2.default.object.isRequired,
  mapFields: _propTypes2.default.object.isRequired
};

PlotContainerFactory.deps = [_mapContainer2.default];

var StyledPlotContainer = _styledComponents2.default.div(_templateObject);

function PlotContainerFactory(MapContainer) {
  var _class, _temp, _initialiseProps;

  var PlotContainer = (_temp = _class = function (_Component) {
    (0, _inherits3.default)(PlotContainer, _Component);

    function PlotContainer(props) {
      (0, _classCallCheck3.default)(this, PlotContainer);

      var _this = (0, _possibleConstructorReturn3.default)(this, (PlotContainer.__proto__ || Object.getPrototypeOf(PlotContainer)).call(this, props));

      _initialiseProps.call(_this);

      _this._onMapRender = (0, _lodash2.default)(_this._onMapRender, 500);
      return _this;
    }

    (0, _createClass3.default)(PlotContainer, [{
      key: 'componentWillMount',
      value: function componentWillMount() {
        this.props.startExportingImage();
      }
    }, {
      key: 'componentWillReceiveProps',
      value: function componentWillReceiveProps(newProps) {
        var _this2 = this;

        // re-fetch the new screenshot only when ratio legend or resolution changes
        var checks = ['ratio', 'resolution', 'legend'];
        var shouldRetrieveScreenshot = checks.some(function (item) {
          return _this2.props.exportImageSetting[item] !== newProps.exportImageSetting[item];
        });
        if (shouldRetrieveScreenshot) {
          this._retrieveNewScreenshot();
        }
      }
    }, {
      key: 'render',
      value: function render() {
        var _this3 = this;

        var _props = this.props,
            width = _props.width,
            height = _props.height,
            exportImageSetting = _props.exportImageSetting,
            mapFields = _props.mapFields;
        var ratio = exportImageSetting.ratio,
            resolution = exportImageSetting.resolution,
            legend = exportImageSetting.legend;

        var exportImageSize = (0, _exportImageUtils.calculateExportImageSize)({
          width: width,
          height: height,
          ratio: ratio,
          resolution: resolution
        });

        var mapProps = (0, _extends3.default)({}, mapFields, {
          mapStyle: this.scaledMapStyleSelector(this.props),

          // override viewport based on export settings
          mapState: (0, _extends3.default)({}, mapFields.mapState, exportImageSize, {
            zoom: mapFields.mapState.zoom + exportImageSize.zoomOffset
          }),
          mapControls: {
            // override map legend visibility
            mapLegend: {
              show: legend,
              active: true
            }
          },
          MapComponent: _reactMapGl.StaticMap
        });

        return _react2.default.createElement(
          StyledPlotContainer,
          {
            style: { position: 'absolute', top: -9999, left: -9999 }
          },
          _react2.default.createElement(
            'div',
            {
              ref: function ref(element) {
                _this3.plottingAreaRef = element;
              },
              style: {
                width: exportImageSize.width,
                height: exportImageSize.height
              }
            },
            _react2.default.createElement(MapContainer, (0, _extends3.default)({
              index: 0,
              onMapRender: this._onMapRender,
              isExport: true
            }, mapProps))
          )
        );
      }
    }]);
    return PlotContainer;
  }(_react.Component), _initialiseProps = function _initialiseProps() {
    var _this4 = this;

    this.mapStyleSelector = function (props) {
      return props.mapFields.mapStyle;
    };

    this.resolutionSelector = function (props) {
      return props.exportImageSetting.resolution;
    };

    this.scaledMapStyleSelector = (0, _reselect.createSelector)(this.mapStyleSelector, this.resolutionSelector, function (mapStyle, resolution) {
      return {
        bottomMapStyle: (0, _mapboxGlStyleEditor.scaleMapStyleByResolution)(mapStyle.bottomMapStyle, resolution),
        topMapStyle: (0, _mapboxGlStyleEditor.scaleMapStyleByResolution)(mapStyle.topMapStyle, resolution)
      };
    });

    this._onMapRender = function (map) {
      if (map.isStyleLoaded()) {
        _this4._retrieveNewScreenshot();
      }
    };

    this._retrieveNewScreenshot = function () {
      if (_this4.plottingAreaRef) {
        // setting windowDevicePixelRatio to 1
        // so that large mapbox base map will load in full
        var savedDevicePixelRatio = _window2.default.devicePixelRatio;
        _window2.default.devicePixelRatio = 1;

        _this4.props.startExportingImage();
        (0, _exportImageUtils.convertToPng)(_this4.plottingAreaRef).then(function (dataUri) {
          _this4.props.setExportImageDataUri({ dataUri: dataUri });
          _window2.default.devicePixelRatio = savedDevicePixelRatio;
        });
      }
    };
  }, _temp);


  PlotContainer.propsTypes = propTypes;
  return PlotContainer;
}
