// kepler.gl/dist/components/side-panel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  ', ';\n  flex-grow: 1;\n  padding: 16px;\n  overflow-y: scroll;\n  overflow-x: hidden;\n'], ['\n  ', ';\n  flex-grow: 1;\n  padding: 16px;\n  overflow-y: scroll;\n  overflow-x: hidden;\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  color: ', ';\n  font-size: 20px;\n  font-weight: 400;\n  letter-spacing: 1.25px;\n  margin-bottom: 14px;\n'], ['\n  color: ', ';\n  font-size: 20px;\n  font-weight: 400;\n  letter-spacing: 1.25px;\n  margin-bottom: 14px;\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

exports.default = SidePanelFactory;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _sideBar = require('./side-panel/side-bar');

var _sideBar2 = _interopRequireDefault(_sideBar);

var _panelHeader = require('./side-panel/panel-header');

var _panelHeader2 = _interopRequireDefault(_panelHeader);

var _layerManager = require('./side-panel/layer-manager');

var _layerManager2 = _interopRequireDefault(_layerManager);

var _filterManager = require('./side-panel/filter-manager');

var _filterManager2 = _interopRequireDefault(_filterManager);

var _interactionManager = require('./side-panel/interaction-manager');

var _interactionManager2 = _interopRequireDefault(_interactionManager);

var _mapManager = require('./side-panel/map-manager');

var _mapManager2 = _interopRequireDefault(_mapManager);

var _panelToggle = require('./side-panel/panel-toggle');

var _panelToggle2 = _interopRequireDefault(_panelToggle);

var _defaultSettings = require('../constants/default-settings');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SidePanelContent = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.sidePanelScrollBar;
});

var PanelTitle = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.theme.titleTextColor;
});

SidePanelFactory.deps = [_panelHeader2.default];

/**
 *
 * Vertical sidebar containing input components for the rendering layers
 */
function SidePanelFactory(PanelHeader) {
  var _class, _temp2;

  return _temp2 = _class = function (_Component) {
    (0, _inherits3.default)(SidePanel, _Component);

    function SidePanel() {
      var _ref;

      var _temp, _this, _ret;

      (0, _classCallCheck3.default)(this, SidePanel);

      for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = SidePanel.__proto__ || Object.getPrototypeOf(SidePanel)).call.apply(_ref, [this].concat(args))), _this), _this._onOpenOrClose = function () {
        _this.props.uiStateActions.toggleSidePanel(_this.props.uiState.activeSidePanel ? null : 'layer');
      }, _this._showDatasetTable = function (dataId) {
        // this will open data table modal
        _this.props.visStateActions.showDatasetTable(dataId);
        _this.props.uiStateActions.toggleModal(_defaultSettings.DATA_TABLE_ID);
      }, _this._showAddDataModal = function () {
        _this.props.uiStateActions.toggleModal(_defaultSettings.ADD_DATA_ID);
      }, _this._showAddMapStyleModal = function () {
        _this.props.uiStateActions.toggleModal(_defaultSettings.ADD_MAP_STYLE_ID);
      }, _this._removeDataset = function (key) {
        // this will show the modal dialog to confirm deletion
        _this.props.uiStateActions.openDeleteModal(key);
      }, _this._onExportImage = function () {
        return _this.props.uiStateActions.toggleModal(_defaultSettings.EXPORT_IMAGE_ID);
      }, _this._onExportData = function () {
        return _this.props.uiStateActions.toggleModal(_defaultSettings.EXPORT_DATA_ID);
      }, _this._onExportConfig = function () {
        return _this.props.uiStateActions.toggleModal(_defaultSettings.EXPORT_CONFIG_ID);
      }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
    }

    /* component private functions */


    (0, _createClass3.default)(SidePanel, [{
      key: 'render',
      value: function render() {
        var _props = this.props,
            appName = _props.appName,
            version = _props.version,
            datasets = _props.datasets,
            filters = _props.filters,
            layers = _props.layers,
            layerBlending = _props.layerBlending,
            layerClasses = _props.layerClasses,
            uiState = _props.uiState,
            layerOrder = _props.layerOrder,
            interactionConfig = _props.interactionConfig,
            visStateActions = _props.visStateActions,
            mapStyleActions = _props.mapStyleActions,
            uiStateActions = _props.uiStateActions;
        var activeSidePanel = uiState.activeSidePanel;

        var isOpen = Boolean(activeSidePanel);

        var layerManagerActions = {
          addLayer: visStateActions.addLayer,
          layerConfigChange: visStateActions.layerConfigChange,
          layerVisualChannelConfigChange: visStateActions.layerVisualChannelConfigChange,
          layerTypeChange: visStateActions.layerTypeChange,
          layerVisConfigChange: visStateActions.layerVisConfigChange,
          updateLayerBlending: visStateActions.updateLayerBlending,
          updateLayerOrder: visStateActions.reorderLayer,
          showDatasetTable: this._showDatasetTable,
          showAddDataModal: this._showAddDataModal,
          removeLayer: visStateActions.removeLayer,
          removeDataset: this._removeDataset
        };

        var filterManagerActions = {
          addFilter: visStateActions.addFilter,
          removeFilter: visStateActions.removeFilter,
          setFilter: visStateActions.setFilter,
          showDatasetTable: this._showDatasetTable,
          showAddDataModal: this._showAddDataModal,
          toggleAnimation: visStateActions.toggleAnimation,
          enlargeFilter: visStateActions.enlargeFilter
        };

        var interactionManagerActions = {
          onConfigChange: visStateActions.interactionConfigChange
        };

        var mapManagerActions = {
          addMapStyleUrl: mapStyleActions.addMapStyleUrl,
          onConfigChange: mapStyleActions.mapConfigChange,
          onStyleChange: mapStyleActions.mapStyleChange,
          onBuildingChange: mapStyleActions.mapBuildingChange,
          showAddMapStyleModal: this._showAddMapStyleModal
        };

        return _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement(
            _sideBar2.default,
            {
              width: this.props.width,
              isOpen: isOpen,
              minifiedWidth: 0,
              onOpenOrClose: this._onOpenOrClose
            },
            _react2.default.createElement(PanelHeader, {
              appName: appName,
              version: version,
              onExportImage: this._onExportImage,
              onExportData: this._onExportData,
              visibleDropdown: uiState.visibleDropdown,
              showExportDropdown: uiStateActions.showExportDropdown,
              hideExportDropdown: uiStateActions.hideExportDropdown,
              onExportConfig: this._onExportConfig,
              onSaveMap: this.props.onSaveMap
            }),
            _react2.default.createElement(_panelToggle2.default, {
              panels: _defaultSettings.PANELS,
              activePanel: activeSidePanel,
              togglePanel: uiStateActions.toggleSidePanel
            }),
            _react2.default.createElement(
              SidePanelContent,
              { className: 'side-panel__content' },
              _react2.default.createElement(
                'div',
                null,
                _react2.default.createElement(
                  PanelTitle,
                  { className: 'side-panel__content__title' },
                  (_defaultSettings.PANELS.find(function (_ref2) {
                    var id = _ref2.id;
                    return id === activeSidePanel;
                  }) || {}).label
                ),
                activeSidePanel === 'layer' && _react2.default.createElement(_layerManager2.default, (0, _extends3.default)({}, layerManagerActions, {
                  datasets: datasets,
                  layers: layers,
                  layerClasses: layerClasses,
                  layerOrder: layerOrder,
                  layerBlending: layerBlending,
                  openModal: uiStateActions.toggleModal
                })),
                activeSidePanel === 'filter' && _react2.default.createElement(_filterManager2.default, (0, _extends3.default)({}, filterManagerActions, {
                  datasets: datasets,
                  filters: filters
                })),
                activeSidePanel === 'interaction' && _react2.default.createElement(_interactionManager2.default, (0, _extends3.default)({}, interactionManagerActions, {
                  datasets: datasets,
                  interactionConfig: interactionConfig
                })),
                activeSidePanel === 'map' && _react2.default.createElement(_mapManager2.default, (0, _extends3.default)({}, mapManagerActions, {
                  mapStyle: this.props.mapStyle
                }))
              )
            )
          )
        );
      }
    }]);
    return SidePanel;
  }(_react.Component), _class.propTypes = {
    filters: _propTypes2.default.arrayOf(_propTypes2.default.any).isRequired,
    interactionConfig: _propTypes2.default.object.isRequired,
    layerBlending: _propTypes2.default.string.isRequired,
    layers: _propTypes2.default.arrayOf(_propTypes2.default.any).isRequired,
    layerClasses: _propTypes2.default.object.isRequired,
    mapStyle: _propTypes2.default.object.isRequired,
    width: _propTypes2.default.number.isRequired,
    datasets: _propTypes2.default.object.isRequired,
    visStateActions: _propTypes2.default.object.isRequired,
    mapStyleActions: _propTypes2.default.object.isRequired
  }, _temp2;
}
