// kepler.gl/dist/reducers/core.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.uiStateLens = exports.visStateLens = exports.mapStyleLens = exports.mapStateLens = exports.coreReducerFactory = undefined;

var _redux = require('redux');

var _visState = require('./vis-state');

var _mapState = require('./map-state');

var _mapStyle = require('./map-style');

var _uiState = require('./ui-state');

var _composers = require('./composers');

var _composers2 = _interopRequireDefault(_composers);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var combined = function combined() {
  var initialState = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return (0, _redux.combineReducers)({
    visState: (0, _visState.visStateReducerFactory)(initialState.visState),
    mapState: (0, _mapState.mapStateReducerFactory)(initialState.mapState),
    mapStyle: (0, _mapStyle.mapStyleReducerFactory)(initialState.mapStyle),
    uiState: (0, _uiState.uiStateReducerFactory)(initialState.uiState)
  });
};

var coreReducerFactory = exports.coreReducerFactory = function coreReducerFactory() {
  var initialState = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  return function (state, action) {
    if (_composers2.default[action.type]) {
      return _composers2.default[action.type](state, action);
    }
    return combined(initialState)(state, action);
  };
};

exports.default = coreReducerFactory();
var mapStateLens = exports.mapStateLens = function mapStateLens(reduxState) {
  return { mapState: reduxState.mapState };
};
var mapStyleLens = exports.mapStyleLens = function mapStyleLens(reduxState) {
  return { mapStyle: reduxState.mapStyle };
};
var visStateLens = exports.visStateLens = function visStateLens(reduxState) {
  return { visState: reduxState.visState };
};
var uiStateLens = exports.uiStateLens = function uiStateLens(reduxState) {
  return { uiState: reduxState.uiState };
};
