// kepler.gl/dist/reducers/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.visStateMergers = exports.combineUpdaters = exports.uiStateUpdaters = exports.mapStyleUpdaters = exports.mapStateUpdaters = exports.visStateUpdaters = exports.mapStyleReducer = exports.mapStateReducer = exports.visStateReducer = exports.mapStyleLens = exports.uiStateLens = exports.mapStateLens = exports.visStateLens = exports.keplerGlReducerCore = exports.keplerGlReducer = exports.default = undefined;

var _root = require('./root');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_root).default;
  }
});
Object.defineProperty(exports, 'keplerGlReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_root).default;
  }
});

var _core = require('./core');

Object.defineProperty(exports, 'keplerGlReducerCore', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_core).default;
  }
});
Object.defineProperty(exports, 'visStateLens', {
  enumerable: true,
  get: function get() {
    return _core.visStateLens;
  }
});
Object.defineProperty(exports, 'mapStateLens', {
  enumerable: true,
  get: function get() {
    return _core.mapStateLens;
  }
});
Object.defineProperty(exports, 'uiStateLens', {
  enumerable: true,
  get: function get() {
    return _core.uiStateLens;
  }
});
Object.defineProperty(exports, 'mapStyleLens', {
  enumerable: true,
  get: function get() {
    return _core.mapStyleLens;
  }
});

var _visState = require('./vis-state');

Object.defineProperty(exports, 'visStateReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_visState).default;
  }
});

var _mapState = require('./map-state');

Object.defineProperty(exports, 'mapStateReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mapState).default;
  }
});

var _mapStyle = require('./map-style');

Object.defineProperty(exports, 'mapStyleReducer', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_mapStyle).default;
  }
});

var _visStateUpdaters2 = require('./vis-state-updaters');

var _visStateUpdaters = _interopRequireWildcard(_visStateUpdaters2);

var _mapStateUpdaters2 = require('./map-state-updaters');

var _mapStateUpdaters = _interopRequireWildcard(_mapStateUpdaters2);

var _mapStyleUpdaters2 = require('./map-style-updaters');

var _mapStyleUpdaters = _interopRequireWildcard(_mapStyleUpdaters2);

var _uiStateUpdaters2 = require('./ui-state-updaters');

var _uiStateUpdaters = _interopRequireWildcard(_uiStateUpdaters2);

var _composers = require('./composers');

var _combineUpdaters = _interopRequireWildcard(_composers);

var _visStateMerger = require('./vis-state-merger');

var _visStateMergers = _interopRequireWildcard(_visStateMerger);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// reducer updaters


exports.visStateUpdaters = _visStateUpdaters;
exports.mapStateUpdaters = _mapStateUpdaters;
exports.mapStyleUpdaters = _mapStyleUpdaters;
exports.uiStateUpdaters = _uiStateUpdaters;
exports.combineUpdaters = _combineUpdaters;
// reducer merges

exports.visStateMergers = _visStateMergers;
