// kepler.gl/dist/reducers/ui-state-updaters.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setExportDataUpdater = exports.setExportConfigUpdater = exports.setExportFilteredUpdater = exports.setExportDataTypeUpdater = exports.setExportSelectedDatasetUpdater = exports.cleanupExportImage = exports.setExportImageDataUri = exports.startExportingImage = exports.setResolutionUpdater = exports.setRatioUpdater = exports.toggleLegendUpdater = exports.openDeleteModalUpdater = exports.toggleMapControlUpdater = exports.hideExportDropdownUpdater = exports.showExportDropdownUpdater = exports.toggleModalUpdater = exports.toggleSidePanelUpdater = undefined;

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends3 = require('babel-runtime/helpers/extends');

var _extends4 = _interopRequireDefault(_extends3);

var _defaultSettings = require('../constants/default-settings');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* Updaters */
var toggleSidePanelUpdater = exports.toggleSidePanelUpdater = function toggleSidePanelUpdater(state, _ref) {
  var id = _ref.payload;

  if (id === state.activeSidePanel) {
    return state;
  }

  if (id === _defaultSettings.LAYER_CONFIG_ID) {
    return (0, _extends4.default)({}, state, {
      currentModal: id
    });
  }

  return (0, _extends4.default)({}, state, {
    activeSidePanel: id
  });
}; // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var toggleModalUpdater = exports.toggleModalUpdater = function toggleModalUpdater(state, _ref2) {
  var id = _ref2.payload;
  return (0, _extends4.default)({}, state, {
    currentModal: id
  });
};

var showExportDropdownUpdater = exports.showExportDropdownUpdater = function showExportDropdownUpdater(state, _ref3) {
  var id = _ref3.payload;
  return (0, _extends4.default)({}, state, {
    visibleDropdown: id
  });
};

var hideExportDropdownUpdater = exports.hideExportDropdownUpdater = function hideExportDropdownUpdater(state, _ref4) {
  var payload = _ref4.payload;
  return (0, _extends4.default)({}, state, {
    visibleDropdown: null
  });
};

var toggleMapControlUpdater = exports.toggleMapControlUpdater = function toggleMapControlUpdater(state, _ref5) {
  var panelId = _ref5.payload;
  return (0, _extends4.default)({}, state, {
    mapControls: (0, _extends4.default)({}, state.mapControls, (0, _defineProperty3.default)({}, panelId, (0, _extends4.default)({}, state.mapControls[panelId], {
      active: !state.mapControls[panelId].active
    })))
  });
};

var openDeleteModalUpdater = exports.openDeleteModalUpdater = function openDeleteModalUpdater(state, _ref6) {
  var datasetKeyToRemove = _ref6.payload;
  return (0, _extends4.default)({}, state, {
    currentModal: _defaultSettings.DELETE_DATA_ID,
    datasetKeyToRemove: datasetKeyToRemove
  });
};

var toggleLegendUpdater = exports.toggleLegendUpdater = function toggleLegendUpdater(state) {
  return (0, _extends4.default)({}, state, {
    exportImage: (0, _extends4.default)({}, state.exportImage, {
      legend: !state.exportImage.legend
    })
  });
};

var setRatioUpdater = exports.setRatioUpdater = function setRatioUpdater(state, _ref7) {
  var payload = _ref7.payload;
  return (0, _extends4.default)({}, state, {
    exportImage: (0, _extends4.default)({}, state.exportImage, {
      ratio: payload.ratio
    })
  });
};

var setResolutionUpdater = exports.setResolutionUpdater = function setResolutionUpdater(state, _ref8) {
  var payload = _ref8.payload;
  return (0, _extends4.default)({}, state, {
    exportImage: (0, _extends4.default)({}, state.exportImage, {
      resolution: payload.resolution
    })
  });
};

var startExportingImage = exports.startExportingImage = function startExportingImage(state) {
  return (0, _extends4.default)({}, state, {
    exportImage: (0, _extends4.default)({}, state.exportImage, {
      exporting: true,
      imageDataUri: ''
    })
  });
};

var setExportImageDataUri = exports.setExportImageDataUri = function setExportImageDataUri(state, _ref9) {
  var payload = _ref9.payload;
  return (0, _extends4.default)({}, state, {
    exportImage: (0, _extends4.default)({}, state.exportImage, {
      exporting: false,
      imageDataUri: payload.dataUri
    })
  });
};

var cleanupExportImage = exports.cleanupExportImage = function cleanupExportImage(state) {
  return (0, _extends4.default)({}, state, {
    exportImage: (0, _extends4.default)({}, state.exportImage, {
      exporting: false,
      imageDataUri: ''
    })
  });
};

var setExportSelectedDatasetUpdater = exports.setExportSelectedDatasetUpdater = function setExportSelectedDatasetUpdater(state, _ref10) {
  var payload = _ref10.payload;
  return (0, _extends4.default)({}, state, {
    exportData: (0, _extends4.default)({}, state.exportData, {
      selectedDataset: payload.dataset
    })
  });
};

var setExportDataTypeUpdater = exports.setExportDataTypeUpdater = function setExportDataTypeUpdater(state, _ref11) {
  var payload = _ref11.payload;
  return (0, _extends4.default)({}, state, {
    exportData: (0, _extends4.default)({}, state.exportData, {
      dataType: payload.dataType
    })
  });
};

var setExportFilteredUpdater = exports.setExportFilteredUpdater = function setExportFilteredUpdater(state, _ref12) {
  var payload = _ref12.payload;
  return (0, _extends4.default)({}, state, {
    exportData: (0, _extends4.default)({}, state.exportData, {
      filtered: payload.filtered
    })
  });
};

var setExportConfigUpdater = exports.setExportConfigUpdater = function setExportConfigUpdater(state, action) {
  return (0, _extends4.default)({}, state, {
    exportData: (0, _extends4.default)({}, state.exportData, {
      config: !state.exportData.config
    })
  });
};

var setExportDataUpdater = exports.setExportDataUpdater = function setExportDataUpdater(state, action) {
  return (0, _extends4.default)({}, state, {
    exportData: (0, _extends4.default)({}, state.exportData, {
      data: !state.exportData.data
    })
  });
};
