// probe.gl/dist/es6/lib/utils/formatters.js
export function formatTime(ms) {
  let formatted;

  if (ms < 10) {
    formatted = `${ms.toFixed(2)}ms`;
  } else if (ms < 100) {
    formatted = `${ms.toFixed(1)}ms`;
  } else if (ms < 1000) {
    formatted = `${ms.toFixed(0)}ms`;
  } else {
    formatted = `${(ms / 1000).toFixed(2)}s`;
  }

  return formatted;
}
export function leftPad(string, length = 8) {
  const padLength = Math.max(length - string.length, 0);
  return `${' '.repeat(padLength)}${string}`;
}
export function rightPad(string, length = 8) {
  const padLength = Math.max(length - string.length, 0);
  return `${string}${' '.repeat(padLength)}`;
}

function splitIntoBaseAndExponent(number) {
  let base = number;
  let exponent = 0;

  if (number !== 0) {
    while (base >= 10 || base <= -10) {
      base /= 10;
      exponent++;
    }

    while (base < 1 && base > -1) {
      base *= 10;
      exponent--;
    }
  }

  return {
    base,
    exponent
  };
}

export function getSISuffix(multipleOf3) {
  const SI_SUFFIXES = {
    0: '',
    1: 'K',
    2: 'M',
    3: 'G',
    '-1': 'm',
    '-2': 'µ',
    '-3': 'n'
  };
  const key = String(multipleOf3);
  return key in SI_SUFFIXES ? SI_SUFFIXES[key] : `e${multipleOf3 * 3}`;
}
export function formatSI(number, precision = 3) {
  const _splitIntoBaseAndExpo = splitIntoBaseAndExponent(number),
        base = _splitIntoBaseAndExpo.base,
        exponent = _splitIntoBaseAndExpo.exponent;

  const multipleOf3 = Math.floor(exponent / 3);
  const remaining = exponent - multipleOf3 * 3;
  const digits = base * Math.pow(10, remaining);
  return `${digits.toPrecision(precision)}${getSISuffix(multipleOf3)}`;
}
export function formatValue(v, opts = {}) {
  const EPSILON = 1e-16;
  const _opts$isInteger = opts.isInteger,
        isInteger = _opts$isInteger === void 0 ? false : _opts$isInteger;

  if (Array.isArray(v) || ArrayBuffer.isView(v)) {
    return formatArrayValue(v, opts);
  }

  if (!Number.isFinite(v)) {
    return String(v);
  }

  if (Math.abs(v) < EPSILON) {
    return isInteger ? '0' : '0.';
  }

  if (isInteger) {
    return v.toFixed(0);
  }

  if (Math.abs(v) > 100 && Math.abs(v) < 10000) {
    return v.toFixed(0);
  }

  const string = v.toPrecision(2);
  const decimal = string.indexOf('.0');
  return decimal === string.length - 2 ? string.slice(0, -1) : string;
}

function formatArrayValue(v, opts) {
  const _opts$maxElts = opts.maxElts,
        maxElts = _opts$maxElts === void 0 ? 16 : _opts$maxElts,
        _opts$size = opts.size,
        size = _opts$size === void 0 ? 1 : _opts$size;
  let string = '[';

  for (let i = 0; i < v.length && i < maxElts; ++i) {
    if (i > 0) {
      string += `,${i % size === 0 ? ' ' : ''}`;
    }

    string += formatValue(v[i], opts);
  }

  const terminator = v.length > maxElts ? '...' : ']';
  return `${string}${terminator}`;
}

export function formatImage(image, message, scale, maxWidth = 600) {
  const imageUrl = image.src.replace(/\(/g, '%28').replace(/\)/g, '%29');

  if (image.width > maxWidth) {
    scale = Math.min(scale, maxWidth / image.width);
  }

  const width = image.width * scale;
  const height = image.height * scale;
  const style = ['font-size:1px;', `padding:${Math.floor(height / 2)}px ${Math.floor(width / 2)}px;`, `line-height:${height}px;`, `background:url(${imageUrl});`, `background-size:${width}px ${height}px;`, 'color:transparent;'].join('');
  return [`${message} %c+`, style];
}
