// luma.gl/dist/es6/webgl-debug/debug-parameters.js
export const GLenum = 'GLenum';
export const GLfloat = 'GLfloat';
export const GLint = 'GLint';
export const GLuint = 'GLint';
export const GLboolean = 'GLboolean';
export const BUFFER_PARAMETERS = {
  [34660]: {
    webgl1: 0
  },
  [34661]: {
    webgl1: 0
  }
};
export const FENCE_SYNC_PARAMETERS = [37138, 37140, 37139, 37141];
export const FRAMEBUFFER_ATTACHMENT_PARAMETERS = [36049, 36048, 36051, 36050, 33296, 36052, 33298, 33299, 33300, 33301, 33302, 33303, 33297];
export const FRAMEBUFFER_STATUS = {
  [36053]: 'Success. Framebuffer is correctly set up',
  [36054]: 'Framebuffer attachment types mismatched or some attachment point not attachment complete',
  [36055]: 'Framebuffer has no attachment',
  [36057]: 'Framebuffer attachments do not have the same size',
  [36061]: 'Framebuffer attachment format not supported or depth and stencil attachments are not same',
  [36182]: 'Framebuffer attachement SAMPLES differs among renderbuffers, or are mixed with textures'
};
export const PROGRAM_PARAMETERS = {
  [35712]: {
    webgl1: 0
  },
  [35714]: {
    webgl1: 0
  },
  [35715]: {
    webgl1: 0
  },
  [35717]: {
    webgl1: 0
  },
  [35721]: {
    webgl1: 0
  },
  [35718]: {
    webgl1: 0
  },
  [35967]: {
    webgl2: 0
  },
  [35971]: {
    webgl2: 0
  },
  [35382]: {
    webgl2: 0
  }
};
export const RENDERBUFFER_PARAMETERS = {
  [36162]: {
    webgl1: 0
  },
  [36163]: {
    webgl1: 0
  },
  [36164]: {
    type: 'GLenum',
    webgl1: 32854
  },
  [36177]: {
    webgl1: 0
  },
  [36178]: {
    webgl1: 0
  },
  [36176]: {
    webgl1: 0
  },
  [36179]: {
    webgl1: 0
  },
  [36180]: {
    webgl1: 0
  },
  [36181]: {
    webgl1: 0
  },
  [36011]: {
    webgl2: 1
  }
};
export const SAMPLER_PARAMETERS = {
  [10240]: {
    webgl2: true
  },
  [10241]: {
    webgl2: true
  },
  [10242]: {
    webgl2: true
  },
  [10243]: {
    webgl2: true
  },
  [32882]: {
    webgl2: true
  },
  [33084]: {
    webgl2: true
  },
  [33085]: {
    webgl2: true
  },
  [34893]: {
    webgl2: true
  },
  [34892]: {
    webgl2: true
  },
  [33082]: {
    webgl2: true
  },
  [33083]: {
    webgl2: true
  }
};
export const TEXTURE_PARAMETERS = {
  [10240]: {
    type: 'GLenum',
    webgl1: 9729
  },
  [10241]: {
    type: 'GLenum',
    webgl1: 9986
  },
  [10242]: {
    type: 'GLenum',
    webgl1: 10497
  },
  [10243]: {
    type: 'GLenum',
    webgl1: 10497
  },
  [32882]: {
    type: 'GLenum',
    webgl2: 10497
  },
  [33084]: {
    webgl2: 0
  },
  [33085]: {
    webgl2: 1000
  },
  [34893]: {
    type: 'GLenum',
    webgl2: 515
  },
  [34892]: {
    type: 'GLenum',
    webgl2: 0
  },
  [33082]: {
    webgl2: -1000
  },
  [33083]: {
    webgl2: 1000
  },
  [4096]: {
    webgl1: 0
  },
  [4097]: {
    webgl1: 0
  }
};
export function installParameterDefinitions() {}
