// kepler.gl/dist/utils/export-image-utils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

exports.calculateExportImageSize = calculateExportImageSize;
exports.convertToPng = convertToPng;
exports.dataURItoBlob = dataURItoBlob;
exports.downloadFile = downloadFile;

var _domToImage = require('./dom-to-image');

var _domToImage2 = _interopRequireDefault(_domToImage);

var _window = require('global/window');

var _defaultSettings = require('../constants/default-settings');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function calculateExportImageSize(_ref) {
  var width = _ref.width,
      height = _ref.height,
      ratio = _ref.ratio,
      resolution = _ref.resolution;

  var resolutionItem = _defaultSettings.RESOLUTION_OPTIONS.find(function (op) {
    return op.id === resolution;
  });

  var _resolutionItem$getSi = resolutionItem.getSize(width, height),
      scaledWidth = _resolutionItem$getSi.width,
      scaledHeight = _resolutionItem$getSi.height;

  var zoomOffset = resolutionItem.zoomOffset,
      scale = resolutionItem.scale;


  return (0, _extends3.default)({
    zoomOffset: zoomOffset,
    scale: scale
  }, _defaultSettings.RATIO_OPTIONS.find(function (op) {
    return op.id === ratio;
  }).getSize(scaledWidth, scaledHeight));
} // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

function convertToPng(sourceElem) {
  return _domToImage2.default.toPng(sourceElem);
}

function dataURItoBlob(dataURI) {
  var binary = (0, _window.atob)(dataURI.split(',')[1]);

  // separate out the mime component
  var mimeString = dataURI.split(',')[0].split(':')[1].split(';')[0];

  // write the bytes of the string to an ArrayBuffer
  var ab = new _window.ArrayBuffer(binary.length);

  // create a view into the buffer
  var ia = new _window.Uint8Array(ab);

  for (var i = 0; i < binary.length; i++) {
    ia[i] = binary.charCodeAt(i);
  }

  return new _window.Blob([ab], { type: mimeString });
}

function downloadFile(fileBlob, filename) {
  var url = _window.URL.createObjectURL(fileBlob);

  var link = _window.document.createElement('a');
  link.setAttribute('href', url);
  link.setAttribute('download', filename);

  _window.document.body.appendChild(link);
  link.click();
  _window.document.body.removeChild(link);
  _window.URL.revokeObjectURL(url);
}
