// kepler.gl/dist/constants/color-ranges.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DefaultColorRange = exports.COLOR_RANGES = undefined;

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _colorbrewer = require('colorbrewer');

var _colorbrewer2 = _interopRequireDefault(_colorbrewer);

var _customColorRanges = require('./custom-color-ranges');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var colorBrewerMap = {
  YlGn: _customColorRanges.SEQ,
  YlGnBu: _customColorRanges.SEQ,
  GnBu: _customColorRanges.SEQ,
  BuGn: _customColorRanges.SEQ,
  PuBuGn: _customColorRanges.SEQ,
  PuBu: _customColorRanges.SEQ,
  BuPu: _customColorRanges.SEQ,
  RdPu: _customColorRanges.SEQ,
  PuRd: _customColorRanges.SEQ,
  OrRd: _customColorRanges.SEQ,
  YlOrRd: _customColorRanges.SEQ,
  YlOrBr: _customColorRanges.SEQ,
  Purples: _customColorRanges.SEQ,
  Blues: _customColorRanges.SEQ,
  Greens: _customColorRanges.SEQ,
  Oranges: _customColorRanges.SEQ,
  Reds: _customColorRanges.SEQ,
  Greys: _customColorRanges.SEQ,
  PuOr: _customColorRanges.DIV,
  BrBG: _customColorRanges.DIV,
  PRGn: _customColorRanges.DIV,
  PiYG: _customColorRanges.DIV,
  RdBu: _customColorRanges.DIV,
  RdGy: _customColorRanges.DIV,
  RdYlBu: _customColorRanges.DIV,
  Spectral: _customColorRanges.DIV,
  RdYlGn: _customColorRanges.DIV,
  Accent: _customColorRanges.QUA,
  Dark2: _customColorRanges.QUA,
  Paired: _customColorRanges.QUA,
  Pastel1: _customColorRanges.QUA,
  Pastel2: _customColorRanges.QUA,
  Set1: _customColorRanges.QUA,
  Set2: _customColorRanges.QUA,
  Set3: _customColorRanges.QUA
};

var colorRanges = [].concat((0, _toConsumableArray3.default)(_customColorRanges.VizColorPalette));

// Add colorbrewer color schemes (Data Science requirement)
// See http://colorbrewer2.org/
function entries(obj) {
  return Object.keys(obj).map(function (k) {
    return [k, obj[k]];
  });
}

var _iteratorNormalCompletion = true;
var _didIteratorError = false;
var _iteratorError = undefined;

try {
  for (var _iterator = entries(_colorbrewer2.default)[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
    var _step$value = (0, _slicedToArray3.default)(_step.value, 2),
        keyName = _step$value[0],
        colorScheme = _step$value[1];

    var _iteratorNormalCompletion2 = true;
    var _didIteratorError2 = false;
    var _iteratorError2 = undefined;

    try {
      for (var _iterator2 = entries(colorScheme)[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
        var _step2$value = (0, _slicedToArray3.default)(_step2.value, 2),
            lenKey = _step2$value[0],
            colors = _step2$value[1];

        colorRanges.push({
          name: 'ColorBrewer ' + keyName + '-' + lenKey,
          type: colorBrewerMap[keyName],
          category: 'ColorBrewer',
          colors: colors
        });
      }
    } catch (err) {
      _didIteratorError2 = true;
      _iteratorError2 = err;
    } finally {
      try {
        if (!_iteratorNormalCompletion2 && _iterator2.return) {
          _iterator2.return();
        }
      } finally {
        if (_didIteratorError2) {
          throw _iteratorError2;
        }
      }
    }
  }
} catch (err) {
  _didIteratorError = true;
  _iteratorError = err;
} finally {
  try {
    if (!_iteratorNormalCompletion && _iterator.return) {
      _iterator.return();
    }
  } finally {
    if (_didIteratorError) {
      throw _iteratorError;
    }
  }
}

var COLOR_RANGES = exports.COLOR_RANGES = colorRanges;

var DefaultColorRange = exports.DefaultColorRange = colorRanges.find(function (_ref) {
  var name = _ref.name;
  return name === 'Global Warming';
});
