// kepler.gl/dist/constants/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defaultSettings = require('./default-settings');

Object.defineProperty(exports, 'DIMENSIONS', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.DIMENSIONS;
  }
});
Object.defineProperty(exports, 'ALL_FIELD_TYPES', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.ALL_FIELD_TYPES;
  }
});
Object.defineProperty(exports, 'FIELD_OPTS', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.FIELD_OPTS;
  }
});
Object.defineProperty(exports, 'FILTER_TYPES', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.FILTER_TYPES;
  }
});
Object.defineProperty(exports, 'GEOJSON_FIELDS', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.GEOJSON_FIELDS;
  }
});
Object.defineProperty(exports, 'ICON_FIELDS', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.ICON_FIELDS;
  }
});
Object.defineProperty(exports, 'TRIP_POINT_FIELDS', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.TRIP_POINT_FIELDS;
  }
});
Object.defineProperty(exports, 'TRIP_ARC_FIELDS', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.TRIP_ARC_FIELDS;
  }
});
Object.defineProperty(exports, 'SCALE_TYPES', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.SCALE_TYPES;
  }
});
Object.defineProperty(exports, 'LAYER_TYPES', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.LAYER_TYPES;
  }
});
Object.defineProperty(exports, 'LAYER_BLENDINGS', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.LAYER_BLENDINGS;
  }
});
Object.defineProperty(exports, 'AGGREGATION_TYPES', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.AGGREGATION_TYPES;
  }
});
Object.defineProperty(exports, 'MAX_DEFAULT_TOOLTIPS', {
  enumerable: true,
  get: function get() {
    return _defaultSettings.MAX_DEFAULT_TOOLTIPS;
  }
});

var _customColorRanges = require('./custom-color-ranges');

Object.defineProperty(exports, 'VizColorPalette', {
  enumerable: true,
  get: function get() {
    return _customColorRanges.VizColorPalette;
  }
});
Object.defineProperty(exports, 'DataVizColors', {
  enumerable: true,
  get: function get() {
    return _customColorRanges.DataVizColors;
  }
});

var _colorRanges = require('./color-ranges');

Object.defineProperty(exports, 'COLOR_RANGES', {
  enumerable: true,
  get: function get() {
    return _colorRanges.COLOR_RANGES;
  }
});
Object.defineProperty(exports, 'DefaultColorRange', {
  enumerable: true,
  get: function get() {
    return _colorRanges.DefaultColorRange;
  }
});
