// kepler.gl/dist/components/side-panel/layer-panel/color-selector.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _typeof2 = require('babel-runtime/helpers/typeof');

var _typeof3 = _interopRequireDefault(_typeof2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp2;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  width: 32px;\n  height: 18px;\n  border-radius: 1px;\n  background-color: ', ';\n'], ['\n  width: 32px;\n  height: 18px;\n  border-radius: 1px;\n  background-color: ', ';\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  ', ';\n  height: ', ';\n'], ['\n  ', ';\n  height: ', ';\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  justify-content: space-between;\n\n  .color-select__input-group {\n    flex-grow: 1;\n  }\n  .color-select__input-group:nth-child(2) {\n    margin-left: 12px;\n  }\n'], ['\n  display: flex;\n  justify-content: space-between;\n\n  .color-select__input-group {\n    flex-grow: 1;\n  }\n  .color-select__input-group:nth-child(2) {\n    margin-left: 12px;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _colorUtils = require('../../../utils/color-utils');

var _singleColorPalette = require('./single-color-palette');

var _singleColorPalette2 = _interopRequireDefault(_singleColorPalette);

var _colorRangeSelector = require('./color-range-selector');

var _colorRangeSelector2 = _interopRequireDefault(_colorRangeSelector);

var _colorPalette = require('./color-palette');

var _colorPalette2 = _interopRequireDefault(_colorPalette);

var _styledComponents3 = require('../../common/styled-components');

var _reactOnclickoutside = require('react-onclickoutside');

var _reactOnclickoutside2 = _interopRequireDefault(_reactOnclickoutside);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ColorBlock = _styledComponents2.default.div(_templateObject, function (props) {
  return 'rgb(' + props.color.slice(0, 3).join(',') + ')';
});

var ColorSelectorInput = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.inputTheme === 'secondary' ? props.theme.secondaryInput : props.theme.input;
}, function (props) {
  return props.theme.inputBoxHeight;
});

var InputBoxContainer = _styledComponents2.default.div(_templateObject3);

var ColorSelector = (_temp2 = _class = function (_Component) {
  (0, _inherits3.default)(ColorSelector, _Component);

  function ColorSelector() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ColorSelector);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ColorSelector.__proto__ || Object.getPrototypeOf(ColorSelector)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      editing: false
    }, _this.handleClickOutside = function (e) {
      if (_this.state.editing !== false) {
        _this.setState({ editing: false });
      }
    }, _this._onSelectColor = function (color, e) {
      e.stopPropagation();
      if (_this.props.colorSets[_this.state.editing]) {
        _this.props.colorSets[_this.state.editing].setColor(color);
      }
    }, _this._showDropdown = function (e, i) {
      e.stopPropagation();
      e.preventDefault();
      _this.setState({ editing: i });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(ColorSelector, [{
    key: 'render',
    value: function render() {
      var _this2 = this;

      var _props = this.props,
          colorSets = _props.colorSets,
          disabled = _props.disabled,
          inputTheme = _props.inputTheme;
      var editing = this.state.editing;

      var currentEditing = colorSets[editing] && (0, _typeof3.default)(colorSets[editing]) === 'object';

      return _react2.default.createElement(
        'div',
        { className: 'color-selector' },
        _react2.default.createElement(
          InputBoxContainer,
          null,
          colorSets.map(function (cSet, i) {
            return _react2.default.createElement(
              'div',
              { className: 'color-select__input-group', key: i },
              cSet.label ? _react2.default.createElement(
                _styledComponents3.PanelLabel,
                null,
                cSet.label
              ) : null,
              _react2.default.createElement(
                ColorSelectorInput,
                {
                  className: 'color-selector__selector',
                  active: editing === i,
                  disabled: disabled,
                  inputTheme: inputTheme,
                  onMouseDown: function onMouseDown(e) {
                    return _this2._showDropdown(e, i);
                  }
                },
                cSet.isRange ? _react2.default.createElement(_colorPalette2.default, { colors: cSet.selectedColor.colors }) : _react2.default.createElement(ColorBlock, {
                  className: 'color-selector__selector__block',
                  color: cSet.selectedColor
                })
              )
            );
          })
        ),
        currentEditing ? _react2.default.createElement(
          _styledComponents3.StyledPanelDropdown,
          { className: 'color-selector__dropdown' },
          colorSets[editing].isRange ? _react2.default.createElement(_colorRangeSelector2.default, {
            selectedColorRange: colorSets[editing].selectedColor,
            onSelectColorRange: this._onSelectColor
          }) : _react2.default.createElement(_singleColorPalette2.default, {
            selectedColor: (0, _colorUtils.rgbToHex)(colorSets[editing].selectedColor),
            onSelectColor: this._onSelectColor
          })
        ) : null
      );
    }
  }]);
  return ColorSelector;
}(_react.Component), _class.propTypes = {
  colorSets: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    selectedColor: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.any), _propTypes2.default.object]),
    setColor: _propTypes2.default.func.isRequired,
    isRange: _propTypes2.default.bool,
    label: _propTypes2.default.string
  })),
  inputTheme: _propTypes2.default.string,
  disabled: _propTypes2.default.bool
}, _class.defaultProps = {
  colorSets: []
}, _temp2);
;

exports.default = (0, _reactOnclickoutside2.default)(ColorSelector);
