// kepler.gl/dist/components/side-panel/layer-panel/layer-panel-header.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  .layer__remove-layer {\n    opacity: 0;\n  }\n  :hover {\n    cursor: pointer;\n    background-color: ', ';\n    \n    .layer__drag-handle {\n      opacity: 1;\n    }\n    \n    .layer__remove-layer {\n      opacity: 1;  \n    }\n    \n    .layer__enable-config {\n      color: white\n    }\n  }\n'], ['\n  .layer__remove-layer {\n    opacity: 0;\n  }\n  :hover {\n    cursor: pointer;\n    background-color: ', ';\n    \n    .layer__drag-handle {\n      opacity: 1;\n    }\n    \n    .layer__remove-layer {\n      opacity: 1;  \n    }\n    \n    .layer__enable-config {\n      color: white\n    }\n  }\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  color: ', ';\n'], ['\n  display: flex;\n  color: ', ';\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n'], ['\n  display: flex;\n']),
    _templateObject4 = (0, _taggedTemplateLiteral3.default)(['\n  margin-left: 12px;\n\n  .layer__title__type {\n    color: ', ';\n    font-size: 10px;\n    line-height: 12px;\n    letter-spacing: 0.37px;\n    text-transform: capitalize;\n  }\n'], ['\n  margin-left: 12px;\n\n  .layer__title__type {\n    color: ', ';\n    font-size: 10px;\n    line-height: 12px;\n    letter-spacing: 0.37px;\n    text-transform: capitalize;\n  }\n']),
    _templateObject5 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  align-items: center;\n  opacity: 0;\n  \n  :hover {\n    cursor: move;\n    color: ', ';\n  }\n'], ['\n  display: flex;\n  align-items: center;\n  opacity: 0;\n  \n  :hover {\n    cursor: move;\n    color: ', ';\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _panelHeaderAction = require('../panel-header-action');

var _panelHeaderAction2 = _interopRequireDefault(_panelHeaderAction);

var _icons = require('../../common/icons');

var _styledComponents3 = require('../../common/styled-components');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  // required
  id: _propTypes2.default.string.isRequired,
  isDragNDropEnabled: _propTypes2.default.bool,
  isVisible: _propTypes2.default.bool.isRequired,
  label: _propTypes2.default.string.isRequired,
  onToggleVisibility: _propTypes2.default.func.isRequired,

  // optional
  className: _propTypes2.default.string,
  idx: _propTypes2.default.number,
  isConfigActive: _propTypes2.default.bool,
  labelRCGColorValues: _propTypes2.default.array,
  onUpdateLayerLabel: _propTypes2.default.func,
  onRemoveLayer: _propTypes2.default.func
};

var defaultProps = {
  isDragNDropEnabled: true,
  showRemoveLayer: true
};

var StyledLayerPanelHeader = _styledComponents3.StyledPanelHeader.extend(_templateObject, function (props) {
  return props.theme.panelBackgroundHover;
});

var HeaderLabelSection = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.theme.textColor;
});

var HeaderActionSection = _styledComponents2.default.div(_templateObject3);

var LayerTitleSection = _styledComponents2.default.div(_templateObject4, function (props) {
  return props.theme.subtextColor;
});

var DragHandle = _styledComponents2.default.div(_templateObject5, function (props) {
  return props.theme.textColorHl;
});

var LayerPanelHeader = function LayerPanelHeader(_ref) {
  var className = _ref.className,
      idx = _ref.idx,
      isConfigActive = _ref.isConfigActive,
      isDragNDropEnabled = _ref.isDragNDropEnabled,
      isVisible = _ref.isVisible,
      label = _ref.label,
      layerId = _ref.layerId,
      layerType = _ref.layerType,
      labelRCGColorValues = _ref.labelRCGColorValues,
      onToggleVisibility = _ref.onToggleVisibility,
      onUpdateLayerLabel = _ref.onUpdateLayerLabel,
      onToggleEnableConfig = _ref.onToggleEnableConfig,
      onRemoveLayer = _ref.onRemoveLayer,
      showRemoveLayer = _ref.showRemoveLayer;
  return _react2.default.createElement(
    StyledLayerPanelHeader,
    {
      className: (0, _classnames2.default)('layer-panel__header', {
        'sort--handle': !isConfigActive
      }),
      active: isConfigActive,
      labelRCGColorValues: labelRCGColorValues,
      onClick: onToggleEnableConfig
    },
    _react2.default.createElement(
      HeaderLabelSection,
      { className: 'layer-panel__header__content' },
      isDragNDropEnabled && _react2.default.createElement(
        DragHandle,
        { className: 'layer__drag-handle' },
        _react2.default.createElement(_icons.VertDots, { height: '20px' })
      ),
      _react2.default.createElement(_panelHeaderAction2.default, {
        className: 'layer__visibility-toggle',
        id: layerId,
        tooltip: isVisible ? 'hide layer' : 'show layer',
        onClick: onToggleVisibility,
        IconComponent: isVisible ? _icons.EyeSeen : _icons.EyeUnseen,
        active: isVisible,
        flush: true
      }),
      _react2.default.createElement(
        LayerTitleSection,
        { className: 'layer__title' },
        _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement(LayerLabelEditor, { label: label, onEdit: onUpdateLayerLabel }),
          _react2.default.createElement(
            'div',
            { className: 'layer__title__type' },
            layerType
          )
        )
      )
    ),
    _react2.default.createElement(
      HeaderActionSection,
      { className: 'layer-panel__header__actions' },
      showRemoveLayer ? _react2.default.createElement(_panelHeaderAction2.default, {
        className: 'layer__remove-layer',
        id: layerId,
        tooltip: 'Remove layer',
        onClick: onRemoveLayer,
        tooltipType: 'error',
        IconComponent: _icons.Trash
      }) : null,
      _react2.default.createElement(_panelHeaderAction2.default, {
        className: 'layer__enable-config',
        id: layerId,
        tooltip: 'Layer settings',
        onClick: onToggleEnableConfig,
        IconComponent: _icons.ArrowDown
      })
    )
  );
};

var LayerLabelEditor = function LayerLabelEditor(_ref2) {
  var label = _ref2.label,
      onEdit = _ref2.onEdit;
  return _react2.default.createElement(_styledComponents3.InlineInput, {
    type: 'text',
    className: 'layer__title__editor',
    value: label,
    onClick: function onClick(e) {
      e.stopPropagation();
    },
    onChange: onEdit,
    id: 'input-layer-label'
  });
};

LayerPanelHeader.propTypes = propTypes;
LayerPanelHeader.defaultProps = defaultProps;

exports.default = LayerPanelHeader;
