// kepler.gl/dist/components/side-panel/layer-panel/layer-type-selector.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  padding-bottom: 12px;\n  padding-right: 12px;\n\n  &.selected {\n    .layer-type-selector__item__icon {\n      border: 1px solid #caf2f4;\n    }\n  }\n\n  :hover,\n  &.selected {\n    cursor: pointer;\n    .layer-type-selector__item__icon {\n      color: ', ';\n    }\n\n    .layer-type-selector__item__label {\n      color: ', ';\n    }\n  }\n'], ['\n  padding-bottom: 12px;\n  padding-right: 12px;\n\n  &.selected {\n    .layer-type-selector__item__icon {\n      border: 1px solid #caf2f4;\n    }\n  }\n\n  :hover,\n  &.selected {\n    cursor: pointer;\n    .layer-type-selector__item__icon {\n      color: ', ';\n    }\n\n    .layer-type-selector__item__label {\n      color: ', ';\n    }\n  }\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  &.list {\n    display: flex;\n    align-items: center;\n\n    .layer-type-selector__item__icon {\n      color: ', ';\n      background-size: ', 'px ', 'px;\n      margin-right: 12px;  \n    }\n  }\n  \n  .layer-type-selector__item__icon {\n    color: ', ';\n    display: flex;\n    background-image: url(', ');\n    background-size: ', 'px ', 'px;\n  }\n\n  .layer-type-selector__item__label {\n    text-transform: capitalize;\n    font-size: 12px;\n    text-align: center;\n    color: ', ';\n  }\n'], ['\n  &.list {\n    display: flex;\n    align-items: center;\n\n    .layer-type-selector__item__icon {\n      color: ', ';\n      background-size: ', 'px ', 'px;\n      margin-right: 12px;  \n    }\n  }\n  \n  .layer-type-selector__item__icon {\n    color: ', ';\n    display: flex;\n    background-image: url(', ');\n    background-size: ', 'px ', 'px;\n  }\n\n  .layer-type-selector__item__label {\n    text-transform: capitalize;\n    font-size: 12px;\n    text-align: center;\n    color: ', ';\n  }\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  ', ';\n  background-color: ', ';\n  border-top: 1px solid ', ';\n  display: flex;\n  flex-wrap: wrap;\n  align-items: flex-start;\n  padding: 12px 0 0 12px;\n'], ['\n  ', ';\n  background-color: ', ';\n  border-top: 1px solid ', ';\n  display: flex;\n  flex-wrap: wrap;\n  align-items: flex-start;\n  padding: 12px 0 0 12px;\n']),
    _templateObject4 = (0, _taggedTemplateLiteral3.default)(['\n  // override item-selector dropdown padding\n  .item-selector .item-selector__dropdown {\n    padding: 4px 10px 4px 2px;\n  }\n'], ['\n  // override item-selector dropdown padding\n  .item-selector .item-selector__dropdown {\n    padding: 4px 10px 4px 2px;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _dropdownList = require('../../common/item-selector/dropdown-list');

var _itemSelector = require('../../common/item-selector/item-selector');

var _itemSelector2 = _interopRequireDefault(_itemSelector);

var _defaultSettings = require('../../../constants/default-settings');

var _styledComponents3 = require('../../common/styled-components');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var ITEM_SIZE = {
  large: 60,
  small: 28
};

var StyledDropdownListItem = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.activeColor;
}, function (props) {
  return props.theme.textColor;
});

var StyledListItem = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.theme.activeColor;
}, ITEM_SIZE.small, ITEM_SIZE.small, function (props) {
  return props.theme.labelColor;
}, _defaultSettings.CLOUDFRONT + '/kepler.gl-layer-icon-bg.png', ITEM_SIZE.large, ITEM_SIZE.large, function (props) {
  return props.theme.labelColor;
});

var DropdownListWrapper = _styledComponents2.default.div(_templateObject3, function (props) {
  return props.theme.dropdownList;
}, function (props) {
  return props.theme.dropdownListBgd;
}, function (props) {
  return props.theme.dropdownListBorderTop;
});

var LayerTypeListItem = function LayerTypeListItem(_ref) {
  var value = _ref.value,
      isTile = _ref.isTile;
  return _react2.default.createElement(
    StyledListItem,
    {
      className: (0, _classnames2.default)('layer-type-selector__item__inner', { list: !isTile })
    },
    _react2.default.createElement(
      'div',
      { className: 'layer-type-selector__item__icon' },
      _react2.default.createElement(value.icon, {
        height: (isTile ? ITEM_SIZE.large : ITEM_SIZE.small) + 'px'
      })
    ),
    _react2.default.createElement(
      'div',
      { className: 'layer-type-selector__item__label' },
      value.label
    )
  );
};

var LayerTypeDropdownList = function LayerTypeDropdownList(props) {
  return _react2.default.createElement(
    DropdownListWrapper,
    { className: _dropdownList.classList.list },
    props.options.map(function (value, i) {
      return _react2.default.createElement(
        StyledDropdownListItem,
        {
          className: (0, _classnames2.default)('layer-type-selector__item', {
            selected: props.selectedItems.find(function (it) {
              return it.id === value.id;
            }),
            hover: props.selectionIndex === i
          }),
          key: value.id + '_' + i,
          onMouseDown: function onMouseDown(e) {
            e.preventDefault();
            props.onOptionSelected(value, e);
          },
          onClick: function onClick(e) {
            e.preventDefault();
            props.onOptionSelected(value, e);
          }
        },
        _react2.default.createElement(props.customListItemComponent, { value: value, isTile: true })
      );
    })
  );
};

var propTypes = {
  layer: _propTypes2.default.object.isRequired,
  onSelect: _propTypes2.default.func.isRequired
};

var StyledLayerTypeSelector = _styledComponents2.default.div(_templateObject4);
var LayerTypeSelector = function LayerTypeSelector(_ref2) {
  var layer = _ref2.layer,
      layerTypeOptions = _ref2.layerTypeOptions,
      onSelect = _ref2.onSelect;
  return _react2.default.createElement(
    _styledComponents3.SidePanelSection,
    null,
    _react2.default.createElement(
      StyledLayerTypeSelector,
      { className: 'layer-config__type' },
      _react2.default.createElement(
        _styledComponents3.PanelLabel,
        null,
        'Layer type'
      ),
      _react2.default.createElement(_itemSelector2.default, {
        selectedItems: layerTypeOptions.find(function (op) {
          return op.id === layer.type;
        }),
        options: layerTypeOptions,
        multiSelect: false,
        placeholder: 'Select A Type',
        onChange: onSelect,
        getOptionValue: function getOptionValue(op) {
          return op.id;
        },
        filterOption: 'label',
        displayOption: function displayOption(op) {
          return op.label;
        },
        DropDownLineItemRenderComponent: LayerTypeListItem,
        DropDownRenderComponent: LayerTypeDropdownList
      })
    )
  );
};

LayerTypeSelector.propTypes = propTypes;

exports.default = LayerTypeSelector;
