// kepler.gl/dist/components/side-panel/layer-panel/layer-config-group.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledConfigGroupHeader = exports.StyledLayerConfigGroup = exports.StyledLayerConfigGroupLabel = undefined;

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  border-left: 2px solid ', ';\n  color: ', ';\n  font-size: 12px;\n  font-weight: 500;\n  line-height: 12px;\n  margin-left: -12px;\n  padding-left: 10px;\n  text-transform: capitalize;\n  letter-spacing: 0.2px;\n'], ['\n  border-left: 2px solid ', ';\n  color: ', ';\n  font-size: 12px;\n  font-weight: 500;\n  line-height: 12px;\n  margin-left: -12px;\n  padding-left: 10px;\n  text-transform: capitalize;\n  letter-spacing: 0.2px;\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  padding-left: 18px;\n  margin-bottom: 18px;\n\n'], ['\n  padding-left: 18px;\n  margin-bottom: 18px;\n\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n  margin-bottom: 12px;\n'], ['\n  display: flex;\n  align-items: center;\n  justify-content: space-between;\n  margin-bottom: 12px;\n']),
    _templateObject4 = (0, _taggedTemplateLiteral3.default)(['\n  &.disabled {\n    opacity: 0.3;\n    pointer-events: none;\n    * {\n      pointer-events: none;\n    }\n  }\n'], ['\n  &.disabled {\n    opacity: 0.3;\n    pointer-events: none;\n    * {\n      pointer-events: none;\n    }\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _switch = require('../../common/switch');

var _switch2 = _interopRequireDefault(_switch);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StyledLayerConfigGroupLabel = exports.StyledLayerConfigGroupLabel = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.labelColor;
}, function (props) {
  return props.theme.textColor;
});

var StyledLayerConfigGroup = exports.StyledLayerConfigGroup = _styledComponents2.default.div(_templateObject2);

var StyledConfigGroupHeader = exports.StyledConfigGroupHeader = _styledComponents2.default.div(_templateObject3);

var ConfigGroupContent = _styledComponents2.default.div(_templateObject4);

var LayerConfigGroup = function LayerConfigGroup(_ref) {
  var group = _ref.group,
      label = _ref.label,
      children = _ref.children,
      property = _ref.property,
      layer = _ref.layer,
      _onChange2 = _ref.onChange;
  return _react2.default.createElement(
    StyledLayerConfigGroup,
    { className: 'layer-config-group' },
    _react2.default.createElement(
      StyledConfigGroupHeader,
      { className: 'layer-config-group__header' },
      _react2.default.createElement(
        StyledLayerConfigGroupLabel,
        { className: 'layer-config-group__label' },
        label
      ),
      property ? _react2.default.createElement(_switch2.default, {
        checked: layer.config.visConfig[property],
        id: layer.id + '-' + property,
        onChange: function onChange() {
          return _onChange2((0, _defineProperty3.default)({}, property, !layer.config.visConfig[property]));
        }
      }) : null
    ),
    _react2.default.createElement(
      ConfigGroupContent,
      {
        className: (0, _classnames2.default)('layer-config-group__content', {
          disabled: property && !layer.config.visConfig[property]
        })
      },
      children
    )
  );
};

exports.default = LayerConfigGroup;
