// kepler.gl/dist/components/side-panel/layer-panel/vis-config-slider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VisConfigSlider = undefined;

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('../../common/styled-components');

var _utils = require('../../../utils/utils');

var _rangeSlider = require('../../common/range-slider');

var _rangeSlider2 = _interopRequireDefault(_rangeSlider);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  layer: _propTypes2.default.object.isRequired,
  property: _propTypes2.default.string.isRequired,
  onChange: _propTypes2.default.func.isRequired,
  label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.bool, _propTypes2.default.func]),
  range: _propTypes2.default.arrayOf(_propTypes2.default.number).isRequired,
  step: _propTypes2.default.number,
  isRanged: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  inputTheme: _propTypes2.default.bool
}; // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var VisConfigSlider = exports.VisConfigSlider = function VisConfigSlider(_ref) {
  var config = _ref.layer.config,
      property = _ref.property,
      label = _ref.label,
      range = _ref.range,
      step = _ref.step,
      isRanged = _ref.isRanged,
      disabled = _ref.disabled,
      _onChange2 = _ref.onChange,
      inputTheme = _ref.inputTheme;
  return _react2.default.createElement(
    _styledComponents.SidePanelSection,
    { disabled: Boolean(disabled) },
    label ? _react2.default.createElement(
      _styledComponents.PanelLabel,
      null,
      typeof label === 'string' ? label : typeof label === 'function' ? label(config) : (0, _utils.capitalizeFirstLetter)(property)
    ) : null,
    _react2.default.createElement(_rangeSlider2.default, {
      range: range,
      value0: isRanged ? config.visConfig[property][0] : range[0],
      value1: isRanged ? config.visConfig[property][1] : config.visConfig[property],
      step: step,
      isRanged: Boolean(isRanged),
      onChange: function onChange(value) {
        return _onChange2((0, _defineProperty3.default)({}, property, isRanged ? value : value[1]));
      },
      inputTheme: inputTheme,
      showInput: true
    })
  );
};

VisConfigSlider.propTypes = propTypes;

exports.default = VisConfigSlider;
