// kepler.gl/dist/components/side-panel/layer-panel/vis-config-switch.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  justify-content: space-between;\n  \n  .vis-config-switch__title {\n    display: flex;\n  }\n'], ['\n  display: flex;\n  justify-content: space-between;\n  \n  .vis-config-switch__title {\n    display: flex;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _infoHelper = require('../../common/info-helper');

var _infoHelper2 = _interopRequireDefault(_infoHelper);

var _switch = require('../../common/switch');

var _switch2 = _interopRequireDefault(_switch);

var _styledComponents3 = require('../../common/styled-components');

var _utils = require('../../../utils/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  layer: _propTypes2.default.object.isRequired,
  property: _propTypes2.default.string.isRequired,
  onChange: _propTypes2.default.func.isRequired,
  label: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.bool, _propTypes2.default.func]),
  description: _propTypes2.default.string,
  disabled: _propTypes2.default.bool
};

var StyledVisConfigSwitch = _styledComponents2.default.div(_templateObject);

var VisConfigSwitch = function VisConfigSwitch(_ref) {
  var _ref$layer = _ref.layer,
      id = _ref$layer.id,
      config = _ref$layer.config,
      property = _ref.property,
      _onChange2 = _ref.onChange,
      label = _ref.label,
      description = _ref.description,
      disabled = _ref.disabled;
  return _react2.default.createElement(
    _styledComponents3.SidePanelSection,
    { disabled: Boolean(disabled) },
    _react2.default.createElement(
      StyledVisConfigSwitch,
      { className: 'vis-config-switch' },
      _react2.default.createElement(
        'div',
        { className: 'vis-config-switch__title' },
        label ? _react2.default.createElement(
          _styledComponents3.PanelLabel,
          null,
          label || (0, _utils.capitalizeFirstLetter)(property)
        ) : null,
        description ? _react2.default.createElement(
          'div',
          null,
          _react2.default.createElement(_infoHelper2.default, { description: description, id: id + '-' + property })
        ) : null
      ),
      _react2.default.createElement(
        'div',
        { className: 'vis-config-switch__switch' },
        _react2.default.createElement(_switch2.default, {
          checked: config.visConfig[property],
          id: id + '-' + property,
          onChange: function onChange() {
            return _onChange2((0, _defineProperty3.default)({}, property, !config.visConfig[property]));
          }
        })
      )
    )
  );
};

VisConfigSwitch.propTypes = propTypes;

exports.default = VisConfigSwitch;
