// kepler.gl/dist/components/side-panel/layer-panel/vis-config-by-field-selector.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _class, _temp2; // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('../../common/styled-components');

var _fieldSelector = require('../../common/field-selector');

var _fieldSelector2 = _interopRequireDefault(_fieldSelector);

var _infoHelper = require('../../common/info-helper');

var _infoHelper2 = _interopRequireDefault(_infoHelper);

var _dimensionScaleSelector = require('./dimension-scale-selector');

var _dimensionScaleSelector2 = _interopRequireDefault(_dimensionScaleSelector);

var _utils = require('../../../utils/utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var VisConfigByFieldSelector = (_temp2 = _class = function (_Component) {
  (0, _inherits3.default)(VisConfigByFieldSelector, _Component);

  function VisConfigByFieldSelector() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, VisConfigByFieldSelector);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = VisConfigByFieldSelector.__proto__ || Object.getPrototypeOf(VisConfigByFieldSelector)).call.apply(_ref, [this].concat(args))), _this), _this._updateVisByField = function (val) {
      _this.props.updateField(val);
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(VisConfigByFieldSelector, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          property = _props.property,
          showScale = _props.showScale,
          selectedField = _props.selectedField,
          description = _props.description,
          _props$scaleOptions = _props.scaleOptions,
          scaleOptions = _props$scaleOptions === undefined ? [] : _props$scaleOptions;


      return _react2.default.createElement(
        _styledComponents.SidePanelSection,
        null,
        _react2.default.createElement(
          _styledComponents.SidePanelSection,
          null,
          _react2.default.createElement(
            _styledComponents.PanelLabelWrapper,
            null,
            _react2.default.createElement(
              _styledComponents.PanelLabel,
              null,
              this.props.label || (0, _utils.capitalizeFirstLetter)(property) + ' based on'
            ),
            description && _react2.default.createElement(_infoHelper2.default, {
              description: description,
              id: this.props.id + '-' + property
            })
          ),
          _react2.default.createElement(_fieldSelector2.default, {
            fields: this.props.fields,
            value: selectedField && selectedField.name,
            placeholder: this.props.placeholder,
            onSelect: this._updateVisByField,
            erasable: true
          })
        ),
        _react2.default.createElement(
          'div',
          null,
          showScale ? _react2.default.createElement(_dimensionScaleSelector2.default, {
            scaleType: this.props.scaleType,
            options: scaleOptions,
            label: property + ' scale',
            onSelect: this.props.updateScale,
            disabled: scaleOptions.length < 2
          }) : null
        )
      );
    }
  }]);
  return VisConfigByFieldSelector;
}(_react.Component), _class.propTypes = {
  channel: _propTypes2.default.string.isRequired,
  fields: _propTypes2.default.arrayOf(_propTypes2.default.any).isRequired,
  id: _propTypes2.default.string.isRequired,
  property: _propTypes2.default.string.isRequired,
  showScale: _propTypes2.default.bool.isRequired,
  updateField: _propTypes2.default.func.isRequired,
  updateScale: _propTypes2.default.func.isRequired,

  // optional
  scaleType: _propTypes2.default.string,
  selectedField: _propTypes2.default.object,
  description: _propTypes2.default.string,
  label: _propTypes2.default.string,
  placeholder: _propTypes2.default.string
}, _temp2);
exports.default = VisConfigByFieldSelector;
;
