// kepler.gl/dist/components/common/item-selector/dropdown-list.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ListItem = exports.classList = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  background-color: ', ';\n  border-top: 1px solid ', ';\n  ', ';\n'], ['\n  background-color: ', ';\n  border-top: 1px solid ', ';\n  ', ';\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var classList = exports.classList = {
  list: 'list-selector',
  listHeader: 'list__header',
  listSection: 'list__section',
  listItem: 'list__item',
  listItemAnchor: 'list__item__anchor'
};

var defaultDisplay = function defaultDisplay(d) {
  return d;
};
var ListItem = exports.ListItem = function ListItem(_ref) {
  var value = _ref.value,
      _ref$displayOption = _ref.displayOption,
      displayOption = _ref$displayOption === undefined ? defaultDisplay : _ref$displayOption;
  return _react2.default.createElement(
    'span',
    { className: classList.listItemAnchor },
    displayOption(value)
  );
};

var DropdownListWrapper = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.dropdownListBgd;
}, function (props) {
  return props.theme.dropdownListBorderTop;
}, function (props) {
  return props.theme.dropdownList;
});

var DropdownList = (_temp = _class = function (_Component) {
  (0, _inherits3.default)(DropdownList, _Component);

  function DropdownList() {
    (0, _classCallCheck3.default)(this, DropdownList);
    return (0, _possibleConstructorReturn3.default)(this, (DropdownList.__proto__ || Object.getPrototypeOf(DropdownList)).apply(this, arguments));
  }

  (0, _createClass3.default)(DropdownList, [{
    key: '_onClick',
    value: function _onClick(result, event) {
      event.preventDefault();
      this.props.onOptionSelected(result, event);
    }
  }, {
    key: 'render',
    value: function render() {
      var _this2 = this;

      var fixedOptions = this.props.fixedOptions;

      var display = this.props.displayOption;

      // Don't render if there are no options to display
      if (!this.props.options.length && this.props.allowCustomValues <= 0) {
        return false;
      }

      var valueOffset = Array.isArray(fixedOptions) ? fixedOptions.length : 0;

      // For some reason onClick is not fired when clicked on an option
      // onMouseDown is used here as a workaround of #205 and other
      return _react2.default.createElement(
        DropdownListWrapper,
        { className: classList.list },
        this.props.customListHeaderComponent ? _react2.default.createElement(
          'div',
          { className: classList.listHeader },
          _react2.default.createElement(this.props.customListHeaderComponent, null)
        ) : null,
        valueOffset > 0 ? _react2.default.createElement(
          'div',
          { className: classList.listSection },
          fixedOptions.map(function (value, i) {
            return _react2.default.createElement(
              'div',
              {
                className: (0, _classnames2.default)(classList.listItem, {
                  hover: _this2.props.selectionIndex === i,
                  fixed: true
                }),
                key: display(value) + '_' + i,
                onMouseDown: function onMouseDown(e) {
                  return _this2._onClick(value, e);
                },
                onClick: function onClick(e) {
                  return _this2._onClick(value, e);
                }
              },
              _react2.default.createElement(_this2.props.customListItemComponent, {
                value: value,
                displayOption: display
              })
            );
          })
        ) : null,
        this.props.options.map(function (value, i) {
          return _react2.default.createElement(
            'div',
            {
              className: (0, _classnames2.default)(classList.listItem, {
                hover: _this2.props.selectionIndex === i + valueOffset
              }),
              key: display(value) + '_' + i,
              onMouseDown: function onMouseDown(e) {
                return _this2._onClick(value, e);
              },
              onClick: function onClick(e) {
                return _this2._onClick(value, e);
              }
            },
            _react2.default.createElement(_this2.props.customListItemComponent, {
              value: value,
              displayOption: display
            })
          );
        })
      );
    }
  }]);
  return DropdownList;
}(_react.Component), _class.propTypes = {
  options: _propTypes2.default.arrayOf(_propTypes2.default.any),
  allowCustomValues: _propTypes2.default.number,
  customClasses: _propTypes2.default.object,
  customValues: _propTypes2.default.arrayOf(_propTypes2.default.any),
  customListItemComponent: _propTypes2.default.oneOfType([_propTypes2.default.element, _propTypes2.default.func]),
  customListHeaderComponent: _propTypes2.default.oneOfType([_propTypes2.default.element, _propTypes2.default.func]),
  selectionIndex: _propTypes2.default.number,
  onOptionSelected: _propTypes2.default.func,
  displayOption: _propTypes2.default.func.isRequired,
  defaultClassNames: _propTypes2.default.bool,
  areResultsTruncated: _propTypes2.default.bool,
  resultsTruncatedMessage: _propTypes2.default.string,
  listItemComponent: _propTypes2.default.func
}, _class.defaultProps = {
  customClasses: {},
  customListItemComponent: ListItem,
  customListHeaderComponent: null,
  allowCustomValues: 0,
  customValues: [],
  displayOption: defaultDisplay,
  onOptionSelected: function onOptionSelected() {},
  defaultClassNames: true,
  selectionIndex: null
}, _temp);
exports.default = DropdownList;
;
