// kepler.gl/dist/components/side-panel/filter-panel/filter-panel.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp2;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  margin-bottom: 12px;\n  border-radius: 1px;\n\n  .filter-panel__filter {\n    margin-top: 24px;\n  }\n'], ['\n  margin-bottom: 12px;\n  border-radius: 1px;\n\n  .filter-panel__filter {\n    margin-top: 24px;\n  }\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  cursor: pointer;\n  padding: 10px 12px;\n'], ['\n  cursor: pointer;\n  padding: 10px 12px;\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  background-color: ', ';\n  padding: 12px;\n'], ['\n  background-color: ', ';\n  padding: 12px;\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _reselect = require('reselect');

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _panelHeaderAction = require('../panel-header-action');

var _panelHeaderAction2 = _interopRequireDefault(_panelHeaderAction);

var _fieldSelector = require('../../common/field-selector');

var _fieldSelector2 = _interopRequireDefault(_fieldSelector);

var _icons = require('../../common/icons');

var _sourceDataSelector = require('../source-data-selector');

var _sourceDataSelector2 = _interopRequireDefault(_sourceDataSelector);

var _styledComponents3 = require('../../common/styled-components');

var _filters = require('../../filters');

var Filters = _interopRequireWildcard(_filters);

var _filterUtils = require('../../../utils/filter-utils');

var _defaultSettings = require('../../../constants/default-settings');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StyledFilterPanel = _styledComponents2.default.div(_templateObject);

var StyledFilterHeader = _styledComponents3.StyledPanelHeader.extend(_templateObject2);

var StyledFilterContent = _styledComponents2.default.div(_templateObject3, function (props) {
  return props.theme.panelBackground;
});

var FilterPanel = (_temp2 = _class = function (_Component) {
  (0, _inherits3.default)(FilterPanel, _Component);

  function FilterPanel() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, FilterPanel);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = FilterPanel.__proto__ || Object.getPrototypeOf(FilterPanel)).call.apply(_ref, [this].concat(args))), _this), _this.fieldsSelector = function (props) {
      return props.filter.dataId && props.datasets[props.filter.dataId].fields || [];
    }, _this.filterSelector = function (props) {
      return props.filters;
    }, _this.nameSelector = function (props) {
      return props.filter.name;
    }, _this.dataIdSelector = function (props) {
      return props.filter.dataId;
    }, _this.availableFieldsSelector = (0, _reselect.createSelector)(_this.fieldsSelector, _this.filterSelector, _this.nameSelector, _this.dataIdSelector, function (fields, filters, name, dataId) {
      return fields.filter(function (f) {
        return f.type && f.type !== _defaultSettings.ALL_FIELD_TYPES.geojson && (f.name === name || !filters.find(function (d) {
          return d.name === f.name && d.dataId === dataId;
        }));
      });
    }), _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  /* selectors */


  // only show current field and field that's not already been used as a filter


  (0, _createClass3.default)(FilterPanel, [{
    key: 'render',
    value: function render() {
      var _props = this.props,
          datasets = _props.datasets,
          enlargeFilter = _props.enlargeFilter,
          filter = _props.filter,
          idx = _props.idx,
          isAnyFilterAnimating = _props.isAnyFilterAnimating,
          removeFilter = _props.removeFilter,
          _setFilter = _props.setFilter,
          toggleAnimation = _props.toggleAnimation;
      var name = filter.name,
          enlarged = filter.enlarged,
          type = filter.type,
          dataId = filter.dataId;

      var FilterComponent = type && Filters[_filterUtils.FILTER_COMPONENTS[type]];
      var allAvailableFields = this.availableFieldsSelector(this.props);

      return _react2.default.createElement(
        StyledFilterPanel,
        { className: 'filter-panel' },
        _react2.default.createElement(
          StyledFilterHeader,
          { className: 'filter-panel__header',
            labelRCGColorValues: datasets[dataId].color },
          _react2.default.createElement(
            'div',
            { style: { flexGrow: 1 } },
            _react2.default.createElement(_fieldSelector2.default, {
              inputTheme: 'secondary',
              fields: allAvailableFields,
              value: name,
              erasable: false,
              onSelect: function onSelect(value) {
                return _setFilter(idx, 'name', value.name);
              }
            })
          ),
          _react2.default.createElement(_panelHeaderAction2.default, {
            id: filter.id,
            tooltip: 'delete',
            tooltipType: 'error',
            onClick: removeFilter,
            hoverColor: 'errorColor',
            IconComponent: _icons.Trash
          }),
          type === _filterUtils.FILTER_TYPES.timeRange && _react2.default.createElement(_panelHeaderAction2.default, {
            id: filter.id,
            onClick: enlargeFilter,
            tooltip: 'Time Playback',
            IconComponent: _icons.Clock,
            active: enlarged
          })
        ),
        _react2.default.createElement(
          StyledFilterContent,
          { className: 'filter-panel__content' },
          Object.keys(datasets).length > 1 && _react2.default.createElement(_sourceDataSelector2.default, {
            inputTheme: 'secondary',
            datasets: datasets,
            disabled: filter.freeze,
            dataId: filter.dataId,
            onSelect: function onSelect(value) {
              return _setFilter(idx, 'dataId', value);
            }
          }),
          type && !enlarged && _react2.default.createElement(
            'div',
            { className: 'filter-panel__filter' },
            _react2.default.createElement(FilterComponent, {
              filter: filter,
              idx: idx,
              isAnyFilterAnimating: isAnyFilterAnimating,
              toggleAnimation: toggleAnimation,
              setFilter: function setFilter(value) {
                return _setFilter(idx, 'value', value);
              }
            })
          )
        )
      );
    }
  }]);
  return FilterPanel;
}(_react.Component), _class.propTypes = {
  idx: _propTypes2.default.number,
  filters: _propTypes2.default.arrayOf(_propTypes2.default.any).isRequired,
  filter: _propTypes2.default.object.isRequired,
  setFilter: _propTypes2.default.func.isRequired,
  removeFilter: _propTypes2.default.func.isRequired,
  enlargeFilter: _propTypes2.default.func.isRequired,
  toggleAnimation: _propTypes2.default.func.isRequired,
  datasets: _propTypes2.default.object,
  showDatasetTable: _propTypes2.default.func,
  isAnyFilterAnimating: _propTypes2.default.bool
}, _temp2);
exports.default = FilterPanel;
