// kepler.gl/dist/schemas/schema.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _defineProperty2 = require('babel-runtime/helpers/defineProperty');

var _defineProperty3 = _interopRequireDefault(_defineProperty2);

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _window = require('global/window');

var _versions = require('./versions');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var Schema = function () {
  function Schema() {
    var _ref = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {},
        version = _ref.version,
        key = _ref.key,
        properties = _ref.properties;

    (0, _classCallCheck3.default)(this, Schema);

    this.version = version;
    this.properties = properties;
    this.key = key;
  }

  (0, _createClass3.default)(Schema, [{
    key: 'loadPropertiesOrApplySchema',
    value: function loadPropertiesOrApplySchema(node, parent, accumulator) {
      return this._getPropertyValueFromSchema('load', node, parent, accumulator);
    }
  }, {
    key: 'savePropertiesOrApplySchema',
    value: function savePropertiesOrApplySchema(node, parent, accumulator) {
      return this._getPropertyValueFromSchema('save', node, parent, accumulator);
    }
  }, {
    key: '_getPropertyValueFromSchema',
    value: function _getPropertyValueFromSchema(operation, node, parent, accumulator) {
      var _this = this;

      var internal = '_' + operation;
      return (0, _defineProperty3.default)({}, this.key, this.properties ? Object.keys(this.properties).reduce(function (accu, key) {
        return (0, _extends3.default)({}, accu, key in node ? _this.properties[key] ? // if it's another schema
        _this.properties[key][operation] ? // call save or load
        _this.properties[key][internal](node[key], node, accu) : {} : (0, _defineProperty3.default)({}, key, node[key]) : {});
      }, {}) : node);
    }
  }, {
    key: '_isCurrentVersion',
    value: function _isCurrentVersion() {
      return this.version === _versions.CURRENT_VERSION;
    }
  }, {
    key: 'outdatedVersionError',
    value: function outdatedVersionError() {
      if (!this._isCurrentVersion()) {
        _window.console.error(this.key + ' ' + this.version + ' is outdated. save should not be called anymore');
      }
    }
  }, {
    key: '_save',
    value: function _save() {
      // make sure nothing is saved to an outdated version
      this.outdatedVersionError();
      return this.save.apply(this, arguments);
    }
  }, {
    key: 'save',
    value: function save(node, parent) {
      var accumulator = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

      return this.savePropertiesOrApplySchema(node, parent, accumulator);
    }
  }, {
    key: '_load',
    value: function _load() {
      return this.load.apply(this, arguments);
    }
  }, {
    key: 'load',
    value: function load(node, parent) {
      var accumulator = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};

      return this.loadPropertiesOrApplySchema(node, parent, accumulator);
    }
  }]);
  return Schema;
}();

exports.default = Schema;
