// kepler.gl/dist/actions/identity-actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.renameEntry = exports.deleteEntry = exports.registerEntry = exports.RENAME_ENTRY = exports.DELETE_ENTRY = exports.REGISTER_ENTRY = undefined;

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _reduxActions = require('redux-actions');

var _defaultSettings = require('../constants/default-settings');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Actions to add and remove entries
// Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var REGISTER_ENTRY = exports.REGISTER_ENTRY = _defaultSettings.ACTION_PREFIX + 'REGISTER_ENTRY';
var DELETE_ENTRY = exports.DELETE_ENTRY = _defaultSettings.ACTION_PREFIX + 'DELETE_ENTRY';
var RENAME_ENTRY = exports.RENAME_ENTRY = _defaultSettings.ACTION_PREFIX + 'RENAME_ENTRY';

var _map = [REGISTER_ENTRY, DELETE_ENTRY, RENAME_ENTRY].map(function (a) {
  return (0, _reduxActions.createAction)(a);
}),
    _map2 = (0, _slicedToArray3.default)(_map, 3);

var registerEntry = _map2[0],
    deleteEntry = _map2[1],
    renameEntry = _map2[2];
exports.registerEntry = registerEntry;
exports.deleteEntry = deleteEntry;
exports.renameEntry = renameEntry;
