// kepler.gl/dist/actions/map-style-actions.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadCustomMapStyle = exports.loadMapStyleErr = exports.loadMapStyles = exports.mapStyleChange = exports.mapConfigChange = exports.inputMapStyle = exports.addCustomMapStyle = undefined;

var _slicedToArray2 = require('babel-runtime/helpers/slicedToArray');

var _slicedToArray3 = _interopRequireDefault(_slicedToArray2);

var _reduxActions = require('redux-actions');

var _actionTypes = require('../constants/action-types');

var _actionTypes2 = _interopRequireDefault(_actionTypes);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var ADD_CUSTOM_MAP_STYLE = _actionTypes2.default.ADD_CUSTOM_MAP_STYLE,
    INPUT_MAP_STYLE = _actionTypes2.default.INPUT_MAP_STYLE,
    MAP_CONFIG_CHANGE = _actionTypes2.default.MAP_CONFIG_CHANGE,
    MAP_STYLE_CHANGE = _actionTypes2.default.MAP_STYLE_CHANGE,
    LOAD_MAP_STYLES = _actionTypes2.default.LOAD_MAP_STYLES,
    LOAD_MAP_STYLE_ERR = _actionTypes2.default.LOAD_MAP_STYLE_ERR,
    LOAD_CUSTOM_MAP_STYLE = _actionTypes2.default.LOAD_CUSTOM_MAP_STYLE;

// second argument of createAction is expected to be payloadCreator or undefined

var _map = [ADD_CUSTOM_MAP_STYLE, INPUT_MAP_STYLE, MAP_CONFIG_CHANGE, LOAD_MAP_STYLES, LOAD_MAP_STYLE_ERR, MAP_STYLE_CHANGE, LOAD_CUSTOM_MAP_STYLE].map(function (a) {
  return (0, _reduxActions.createAction)(a);
}),
    _map2 = (0, _slicedToArray3.default)(_map, 7),
    addCustomMapStyle = _map2[0],
    inputMapStyle = _map2[1],
    mapConfigChange = _map2[2],
    loadMapStyles = _map2[3],
    loadMapStyleErr = _map2[4],
    mapStyleChange = _map2[5],
    loadCustomMapStyle = _map2[6];

exports.addCustomMapStyle = addCustomMapStyle;
exports.inputMapStyle = inputMapStyle;
exports.mapConfigChange = mapConfigChange;
exports.mapStyleChange = mapStyleChange;
exports.loadMapStyles = loadMapStyles;
exports.loadMapStyleErr = loadMapStyleErr;
exports.loadCustomMapStyle = loadCustomMapStyle;
