// kepler.gl/dist/components/modals/add-map-style-modal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _class, _temp2;

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  flex-direction: column;\n  justify-content: space-around;\n  font-size: 12px;\n  .modal-section {\n    margin-bottom: 32px;\n  }\n  .modal-section:first-child {\n    margin-top: 24px;\n  }\n  \n  .modal-section {\n    .modal-section-title {\n      font-weight: 500;\n    }\n    .modal-section-subtitle {\n      color: ', ';\n    }\n    \n    input {\n      margin-top: 8px;\n    }\n  }\n\n  input {\n    margin-right: 8px;\n  }\n'], ['\n  display: flex;\n  flex-direction: column;\n  justify-content: space-around;\n  font-size: 12px;\n  .modal-section {\n    margin-bottom: 32px;\n  }\n  .modal-section:first-child {\n    margin-top: 24px;\n  }\n  \n  .modal-section {\n    .modal-section-title {\n      font-weight: 500;\n    }\n    .modal-section-subtitle {\n      color: ', ';\n    }\n    \n    input {\n      margin-top: 8px;\n    }\n  }\n\n  input {\n    margin-right: 8px;\n  }\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  align-items: center;\n  display: flex;\n  flex-direction: column;\n  justify-content: center;\n  margin-left: 116px;\n  flex-shrink: 0;\n  width: ', 'px;\n\n  .preview-title {\n    font-weight: 500;\n    font-size: 10px;\n    padding: 8px 0px;\n  }\n  \n  .preview-title.error {\n    color: ', ';\n  }\n\n  .preview-image {\n    background: ', ';\n    border-radius: 4px;\n    box-shadow: 0 8px 16px 0 rgba(0,0,0,0.18);\n    width: ', 'px;\n    height: ', 'px;\n    position: relative;\n  }\n\n  .preview-image-placeholder {\n    position: absolute;\n    top: 0;\n    left: 0;\n  }\n\n  .preview-image-spinner {\n    position: absolute;\n    left: calc(50% - 25px);\n    top: calc(50% - 25px);\n  }\n'], ['\n  align-items: center;\n  display: flex;\n  flex-direction: column;\n  justify-content: center;\n  margin-left: 116px;\n  flex-shrink: 0;\n  width: ', 'px;\n\n  .preview-title {\n    font-weight: 500;\n    font-size: 10px;\n    padding: 8px 0px;\n  }\n  \n  .preview-title.error {\n    color: ', ';\n  }\n\n  .preview-image {\n    background: ', ';\n    border-radius: 4px;\n    box-shadow: 0 8px 16px 0 rgba(0,0,0,0.18);\n    width: ', 'px;\n    height: ', 'px;\n    position: relative;\n  }\n\n  .preview-image-placeholder {\n    position: absolute;\n    top: 0;\n    left: 0;\n  }\n\n  .preview-image-spinner {\n    position: absolute;\n    left: calc(50% - 25px);\n    top: calc(50% - 25px);\n  }\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  font-weight: 500;\n  \n  :hover {\n    cursor: pointer;\n  }\n'], ['\n  font-weight: 500;\n  \n  :hover {\n    cursor: pointer;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

// Utils


var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _reactMapGl = require('react-map-gl');

var _reactMapGl2 = _interopRequireDefault(_reactMapGl);

var _reactDom = require('react-dom');

var _styledComponents3 = require('../common/styled-components');

var _mapboxUtils = require('../../utils/map-style-utils/mapbox-utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var MapH = 190;
var MapW = 264;
var ErrorMsg = {
  styleError: 'Failed to load map style, make sure it is published. For private style, paste in your access token.'
};

var InstructionPanel = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.subtextColorLT;
});

var PreviewMap = _styledComponents2.default.div(_templateObject2, MapW, function (props) {
  return props.theme.errorColor;
}, function (props) {
  return props.theme.modalImagePlaceHolder;
}, MapW, MapH);

var InlineLink = _styledComponents2.default.a(_templateObject3);

var AddMapStyleModal = (_temp2 = _class = function (_Component) {
  (0, _inherits3.default)(AddMapStyleModal, _Component);

  function AddMapStyleModal() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, AddMapStyleModal);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = AddMapStyleModal.__proto__ || Object.getPrototypeOf(AddMapStyleModal)).call.apply(_ref, [this].concat(args))), _this), _this.state = {
      reRenderKey: 0
    }, _this.loadMapStyleJson = function (style) {
      _this.props.loadCustomMapStyle({ style: style, error: false });
    }, _this.loadMapStyleIcon = function () {
      if (_this.mapRef) {
        var canvas = (0, _reactDom.findDOMNode)(_this.mapRef).querySelector('.mapboxgl-canvas');
        var dataUri = canvas.toDataURL();
        _this.props.loadCustomMapStyle({
          icon: dataUri
        });
      }
    }, _this.loadMaoStyleError = function () {
      _this.props.loadCustomMapStyle({ error: true });
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(AddMapStyleModal, [{
    key: 'componentWillReceiveProps',
    value: function componentWillReceiveProps(nextProps) {
      if (this.props.inputStyle.accessToken !== nextProps.inputStyle.accessToken) {
        // toke has changed
        // ReactMapGl doesn't re-create map when token has changed
        // here we force the map to update
        this.setState({
          reRenderKey: this.state.reRenderKey + 1
        });
      }
    }
  }, {
    key: 'componentDidUpdate',
    value: function componentDidUpdate() {
      var _this2 = this;

      var map = this.mapRef && this.mapRef.getMap();
      if (map && this._map !== map) {
        this._map = map;

        map.on('style.load', function () {
          var style = map.getStyle();
          _this2.loadMapStyleJson(style);
        });

        map.on('render', function () {
          if (map.isStyleLoaded()) {
            _this2.loadMapStyleIcon();
          }
        });

        map.on('error', function () {
          _this2.loadMaoStyleError();
        });
      }
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var _props = this.props,
          inputStyle = _props.inputStyle,
          mapState = _props.mapState;


      var mapProps = (0, _extends3.default)({}, mapState, {
        preserveDrawingBuffer: true,
        mapboxApiAccessToken: inputStyle.accessToken || this.props.mapboxApiAccessToken,
        transformRequest: _mapboxUtils.transformRequest
      });

      return _react2.default.createElement(
        'div',
        { className: 'add-map-style-modal' },
        _react2.default.createElement(
          _styledComponents3.StyledModalContent,
          null,
          _react2.default.createElement(
            InstructionPanel,
            null,
            _react2.default.createElement(
              'div',
              { className: 'modal-section' },
              _react2.default.createElement(
                'div',
                { className: 'modal-section-title' },
                '1. Publish your style at mapbox or provide access token'
              ),
              _react2.default.createElement(
                'div',
                { className: 'modal-section-subtitle' },
                'You can create your own map style at',
                _react2.default.createElement(
                  InlineLink,
                  { target: '_blank', href: 'https://www.mapbox.com/studio/styles/' },
                  ' mapbox'
                ),
                ' and',
                _react2.default.createElement(
                  InlineLink,
                  { target: '_blank', href: 'https://www.mapbox.com/help/studio-manual-publish/' },
                  ' publish'
                ),
                ' it.'
              ),
              _react2.default.createElement(
                'div',
                { className: 'modal-section-subtitle' },
                'To use private style, paste your',
                _react2.default.createElement(
                  InlineLink,
                  { target: '_blank', href: 'https://www.mapbox.com/help/how-access-tokens-work/' },
                  ' access token'
                ),
                ' here. *kepler.gl is a client-side application, data stays in your browser..'
              ),
              _react2.default.createElement(_styledComponents3.InputLight, {
                type: 'text',
                value: inputStyle.accessToken || '',
                onChange: function onChange(_ref2) {
                  var value = _ref2.target.value;
                  return _this3.props.inputMapStyle((0, _extends3.default)({}, inputStyle, { accessToken: value }));
                },
                placeholder: 'e.g. pk.abcdefg.xxxxxx'
              })
            ),
            _react2.default.createElement(
              'div',
              { className: 'modal-section' },
              _react2.default.createElement(
                'div',
                { className: 'modal-section-title' },
                '2. Paste style url'
              ),
              _react2.default.createElement(
                'div',
                { className: 'modal-section-subtitle' },
                'What is a',
                _react2.default.createElement(
                  InlineLink,
                  { target: '_blank', href: 'https://www.mapbox.com/help/studio-manual-publish/#style-url' },
                  ' style URL'
                )
              ),
              _react2.default.createElement(_styledComponents3.InputLight, {
                type: 'text',
                value: inputStyle.url || '',
                onChange: function onChange(_ref3) {
                  var value = _ref3.target.value;
                  return _this3.props.inputMapStyle((0, _extends3.default)({}, inputStyle, { url: value }));
                },
                placeholder: 'e.g. mapbox://styles/uberdataviz/abcdefghijklmnopq'
              })
            ),
            _react2.default.createElement(
              'div',
              { className: 'modal-section' },
              _react2.default.createElement(
                'div',
                { className: 'modal-section-title' },
                '3. Name your style'
              ),
              _react2.default.createElement(_styledComponents3.InputLight, {
                type: 'text',
                value: inputStyle.label || '',
                onChange: function onChange(_ref4) {
                  var value = _ref4.target.value;
                  return _this3.props.inputMapStyle((0, _extends3.default)({}, inputStyle, { label: value }));
                }
              })
            )
          ),
          _react2.default.createElement(
            PreviewMap,
            null,
            _react2.default.createElement(
              'div',
              { className: (0, _classnames2.default)('preview-title', { error: inputStyle.error }) },
              inputStyle.error ? ErrorMsg.styleError : inputStyle.style && inputStyle.style.name || ''
            ),
            _react2.default.createElement(
              'div',
              { className: 'preview-image' },
              !inputStyle.isValid ? _react2.default.createElement('div', { className: 'preview-image-spinner' }) : _react2.default.createElement(
                _styledComponents3.StyledMapContainer,
                null,
                _react2.default.createElement(_reactMapGl2.default, (0, _extends3.default)({}, mapProps, {
                  ref: function ref(el) {
                    _this3.mapRef = el;
                  },
                  key: this.state.reRenderKey,
                  width: MapW,
                  height: MapH,
                  mapStyle: inputStyle.url }))
              )
            )
          )
        )
      );
    }
  }]);
  return AddMapStyleModal;
}(_react.Component), _class.propTypes = {
  mapState: _propTypes2.default.object.isRequired,
  inputMapStyle: _propTypes2.default.func.isRequired,
  loadCustomMapStyle: _propTypes2.default.func.isRequired,
  inputStyle: _propTypes2.default.object.isRequired
}, _temp2);


var AddMapStyleModalFactory = function AddMapStyleModalFactory() {
  return AddMapStyleModal;
};
exports.default = AddMapStyleModalFactory;
