// kepler.gl/dist/components/modals/export-config-modal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  flex-direction: row;\n  margin: 35px 0;\n  width: 100%;\n\n  .description {\n    width: 185px;\n    \n    .title {\n      font-weight: 500;\n      color: ', ';\n      font-size: 12px;\n    }\n    .subtitle {\n      color: ', ';\n      font-size: 11px;\n    }\n    \n    .note {\n      color: ', ';\n      font-size: 11px;\n    }\n  }\n  \n  .selection {\n    display: flex;\n    flex-wrap: wrap;\n    flex: 1;\n    padding-left: 50px;\n\n    select {\n      background-color: white;\n      border-radius: 1px;\n      display: inline-block;\n      font: inherit;\n      line-height: 1.5em;\n      padding: 0.5em 3.5em 0.5em 1em;\n      margin: 0;      \n      box-sizing: border-box;\n      appearance: none;\n      width: 250px;\n      height: 36px;\n\n      background-image:\n        linear-gradient(45deg, transparent 50%, gray 50%),\n        linear-gradient(135deg, gray 50%, transparent 50%),\n        linear-gradient(to right, #ccc, #ccc);\n      background-position:\n        calc(100% - 20px) calc(1em + 2px),\n        calc(100% - 15px) calc(1em + 2px),\n        calc(100% - 2.5em) 4.5em;\n      background-size:\n        5px 5px,\n        5px 5px,\n        1px 1.5em;\n      background-repeat: no-repeat;\n    }\n\n    select:focus {\n      background-image:\n        linear-gradient(45deg, green 50%, transparent 50%),\n        linear-gradient(135deg, transparent 50%, green 50%),\n        linear-gradient(to right, #ccc, #ccc);\n      background-position:\n        calc(100% - 15px) 1em,\n        calc(100% - 20px) 1em,\n        calc(100% - 2.5em) 4.5em;\n      background-size:\n        5px 5px,\n        5px 5px,\n        1px 1.5em;\n      background-repeat: no-repeat;\n      border-color: green;\n      outline: 0;\n    }\n  }\n'], ['\n  display: flex;\n  flex-direction: row;\n  margin: 35px 0;\n  width: 100%;\n\n  .description {\n    width: 185px;\n    \n    .title {\n      font-weight: 500;\n      color: ', ';\n      font-size: 12px;\n    }\n    .subtitle {\n      color: ', ';\n      font-size: 11px;\n    }\n    \n    .note {\n      color: ', ';\n      font-size: 11px;\n    }\n  }\n  \n  .selection {\n    display: flex;\n    flex-wrap: wrap;\n    flex: 1;\n    padding-left: 50px;\n\n    select {\n      background-color: white;\n      border-radius: 1px;\n      display: inline-block;\n      font: inherit;\n      line-height: 1.5em;\n      padding: 0.5em 3.5em 0.5em 1em;\n      margin: 0;      \n      box-sizing: border-box;\n      appearance: none;\n      width: 250px;\n      height: 36px;\n\n      background-image:\n        linear-gradient(45deg, transparent 50%, gray 50%),\n        linear-gradient(135deg, gray 50%, transparent 50%),\n        linear-gradient(to right, #ccc, #ccc);\n      background-position:\n        calc(100% - 20px) calc(1em + 2px),\n        calc(100% - 15px) calc(1em + 2px),\n        calc(100% - 2.5em) 4.5em;\n      background-size:\n        5px 5px,\n        5px 5px,\n        1px 1.5em;\n      background-repeat: no-repeat;\n    }\n\n    select:focus {\n      background-image:\n        linear-gradient(45deg, green 50%, transparent 50%),\n        linear-gradient(135deg, transparent 50%, green 50%),\n        linear-gradient(to right, #ccc, #ccc);\n      background-position:\n        calc(100% - 15px) 1em,\n        calc(100% - 20px) 1em,\n        calc(100% - 2.5em) 4.5em;\n      background-size:\n        5px 5px,\n        5px 5px,\n        1px 1.5em;\n      background-repeat: no-repeat;\n      border-color: green;\n      outline: 0;\n    }\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _reactAce = require('react-ace');

var _reactAce2 = _interopRequireDefault(_reactAce);

var _switch = require('../common/switch');

var _switch2 = _interopRequireDefault(_switch);

var _styledComponents3 = require('../common/styled-components');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var propTypes = {
  config: _propTypes2.default.object.required
};

var StyledExportConfigSection = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.textColorLT;
}, function (props) {
  return props.theme.textColor;
}, function (props) {
  return props.theme.errorColor;
});

var ExportConfigModal = function ExportConfigModal(_ref) {
  var data = _ref.data,
      config = _ref.config,
      onChangeExportData = _ref.onChangeExportData;
  return _react2.default.createElement(
    'div',
    { className: 'export-config-modal' },
    _react2.default.createElement(
      _styledComponents3.StyledModalContent,
      null,
      _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          StyledExportConfigSection,
          null,
          _react2.default.createElement(
            'div',
            { className: 'description' },
            _react2.default.createElement(
              'div',
              { className: 'title' },
              'Current Config'
            ),
            _react2.default.createElement(
              'div',
              { className: 'subtitle' },
              'You can copy or export the current Kepler.gl configuration.'
            ),
            _react2.default.createElement(
              'div',
              { className: 'note' },
              '* kepler.gl map config is coupled with loaded datasets. dataId key is used to bind layers and filters to a specific dataset. If you try to upload a configuration with a specific dataId you also need to make sure you existing dataset id match the dataId/s in the config.'
            )
          ),
          _react2.default.createElement(
            'div',
            { className: 'selection' },
            _react2.default.createElement(_reactAce2.default, {
              mode: 'json',
              theme: 'monokai',
              name: 'kepler.gl-config',
              fontSize: 9,
              readOnly: true,
              wrapEnabled: true,
              height: '280px',
              showPrintMargin: true,
              showGutter: true,
              highlightActiveLine: false,
              value: JSON.stringify(config, null, 2),
              setOptions: {
                enableBasicAutocompletion: false,
                enableLiveAutocompletion: true,
                enableSnippets: false,
                showLineNumbers: true,
                tabSize: 2
              } })
          )
        ),
        _react2.default.createElement(
          StyledExportConfigSection,
          null,
          _react2.default.createElement(
            'div',
            { className: 'description' },
            _react2.default.createElement(
              'div',
              { className: 'title' },
              'Export Current Map'
            ),
            _react2.default.createElement(
              'div',
              { className: 'subtitle' },
              'Export current map, including data and config. You can later load the same map by loading this file to kepler.gl.'
            )
          ),
          _react2.default.createElement(
            'div',
            { className: 'selection' },
            _react2.default.createElement(_switch2.default, { type: 'checkbox',
              id: 'export-map-config',
              checked: data,
              onChange: onChangeExportData })
          )
        )
      )
    )
  );
};

ExportConfigModal.propTypes = propTypes;

var ExportConfigModalFactory = function ExportConfigModalFactory() {
  return ExportConfigModal;
};
exports.default = ExportConfigModalFactory;
