// kepler.gl/dist/components/modals/export-data-modal.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _taggedTemplateLiteral2 = require('babel-runtime/helpers/taggedTemplateLiteral');

var _taggedTemplateLiteral3 = _interopRequireDefault(_taggedTemplateLiteral2);

var _templateObject = (0, _taggedTemplateLiteral3.default)(['\n  display: flex;\n  flex-direction: row;\n  margin: 35px 0;\n  width: 100%;\n\n  .description {\n    width: 185px;\n    \n    .title {\n      font-weight: 500;\n      color: ', ';\n      font-size: 12px;\n    }\n    .subtitle {\n      color: ', ';\n      font-size: 11px;\n    }\n  }\n\n  .selection {\n    display: flex;\n    flex-wrap: wrap;\n    flex: 1;\n    padding-left: 50px;\n\n    select {\n      background-color: white;\n      border-radius: 1px;\n      display: inline-block;\n      font: inherit;\n      line-height: 1.5em;\n      padding: 0.5em 3.5em 0.5em 1em;\n      margin: 0;      \n      box-sizing: border-box;\n      appearance: none;\n      width: 250px;\n      height: 36px;\n\n      background-image:\n        linear-gradient(45deg, transparent 50%, gray 50%),\n        linear-gradient(135deg, gray 50%, transparent 50%),\n        linear-gradient(to right, #ccc, #ccc);\n      background-position:\n        calc(100% - 20px) calc(1em + 2px),\n        calc(100% - 15px) calc(1em + 2px),\n        calc(100% - 2.5em) 4.5em;\n      background-size:\n        5px 5px,\n        5px 5px,\n        1px 1.5em;\n      background-repeat: no-repeat;\n    }\n\n    select:focus {\n      background-image:\n        linear-gradient(45deg, green 50%, transparent 50%),\n        linear-gradient(135deg, transparent 50%, green 50%),\n        linear-gradient(to right, #ccc, #ccc);\n      background-position:\n        calc(100% - 15px) 1em,\n        calc(100% - 20px) 1em,\n        calc(100% - 2.5em) 4.5em;\n      background-size:\n        5px 5px,\n        5px 5px,\n        1px 1.5em;\n      background-repeat: no-repeat;\n      border-color: green;\n      outline: 0;\n    }\n  }\n'], ['\n  display: flex;\n  flex-direction: row;\n  margin: 35px 0;\n  width: 100%;\n\n  .description {\n    width: 185px;\n    \n    .title {\n      font-weight: 500;\n      color: ', ';\n      font-size: 12px;\n    }\n    .subtitle {\n      color: ', ';\n      font-size: 11px;\n    }\n  }\n\n  .selection {\n    display: flex;\n    flex-wrap: wrap;\n    flex: 1;\n    padding-left: 50px;\n\n    select {\n      background-color: white;\n      border-radius: 1px;\n      display: inline-block;\n      font: inherit;\n      line-height: 1.5em;\n      padding: 0.5em 3.5em 0.5em 1em;\n      margin: 0;      \n      box-sizing: border-box;\n      appearance: none;\n      width: 250px;\n      height: 36px;\n\n      background-image:\n        linear-gradient(45deg, transparent 50%, gray 50%),\n        linear-gradient(135deg, gray 50%, transparent 50%),\n        linear-gradient(to right, #ccc, #ccc);\n      background-position:\n        calc(100% - 20px) calc(1em + 2px),\n        calc(100% - 15px) calc(1em + 2px),\n        calc(100% - 2.5em) 4.5em;\n      background-size:\n        5px 5px,\n        5px 5px,\n        1px 1.5em;\n      background-repeat: no-repeat;\n    }\n\n    select:focus {\n      background-image:\n        linear-gradient(45deg, green 50%, transparent 50%),\n        linear-gradient(135deg, transparent 50%, green 50%),\n        linear-gradient(to right, #ccc, #ccc);\n      background-position:\n        calc(100% - 15px) 1em,\n        calc(100% - 20px) 1em,\n        calc(100% - 2.5em) 4.5em;\n      background-size:\n        5px 5px,\n        5px 5px,\n        1px 1.5em;\n      background-repeat: no-repeat;\n      border-color: green;\n      outline: 0;\n    }\n  }\n']),
    _templateObject2 = (0, _taggedTemplateLiteral3.default)(['\n  border-radius: 2px;\n  border: 1px solid ', ';\n  color: ', ';\n  cursor: pointer;\n  font-weight: 500;\n  height: 100px;\n  margin: 4px;\n  padding: 6px 10px;\n  width: 100px;\n\n  :hover {\n    color: ', ';\n    border: 1px solid ', ';\n  }\n'], ['\n  border-radius: 2px;\n  border: 1px solid ', ';\n  color: ', ';\n  cursor: pointer;\n  font-weight: 500;\n  height: 100px;\n  margin: 4px;\n  padding: 6px 10px;\n  width: 100px;\n\n  :hover {\n    color: ', ';\n    border: 1px solid ', ';\n  }\n']),
    _templateObject3 = (0, _taggedTemplateLiteral3.default)(['\n  align-items: center;\n  border-radius: 2px;\n  border: 1px solid ', ';\n  cursor: pointer;\n  display: flex;\n  flex-direction: column;\n  height: 60px;\n  justify-content: center;\n  margin: 4px;\n  padding: 8px 12px;\n  width: 140px;\n\n  :hover {\n    border: 1px solid ', ';\n  }\n\n  .filtered-title {\n    color: ', ';\n    font-size: 12px;\n    font-weight: 500;\n  }\n  .filtered-subtitle {\n    color: ', ';\n    font-size: 11px;\n  }\n'], ['\n  align-items: center;\n  border-radius: 2px;\n  border: 1px solid ', ';\n  cursor: pointer;\n  display: flex;\n  flex-direction: column;\n  height: 60px;\n  justify-content: center;\n  margin: 4px;\n  padding: 8px 12px;\n  width: 140px;\n\n  :hover {\n    border: 1px solid ', ';\n  }\n\n  .filtered-title {\n    color: ', ';\n    font-size: 12px;\n    font-weight: 500;\n  }\n  .filtered-subtitle {\n    color: ', ';\n    font-size: 11px;\n  }\n']); // Copyright (c) 2018 Uber Technologies, Inc.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _styledComponents = require('styled-components');

var _styledComponents2 = _interopRequireDefault(_styledComponents);

var _defaultSettings = require('../../constants/default-settings');

var _icons = require('../common/icons');

var _styledComponents3 = require('../common/styled-components');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StyledExportDataSection = _styledComponents2.default.div(_templateObject, function (props) {
  return props.theme.textColorLT;
}, function (props) {
  return props.theme.textColor;
});

var StyledDataType = _styledComponents2.default.div(_templateObject2, function (props) {
  return props.selected ? props.theme.primaryBtnBgd : props.theme.selectBorderColorLT;
}, function (props) {
  return props.selected ? props.theme.primaryBtnBgd : props.theme.selectBorderColorLT;
}, function (props) {
  return props.available && props.theme.primaryBtnBgd;
}, function (props) {
  return props.available && props.theme.primaryBtnBgd;
});

var StyledFilteredDataOption = _styledComponents2.default.div(_templateObject3, function (props) {
  return props.selected ? props.theme.primaryBtnBgd : props.theme.selectBorderColorLT;
}, function (props) {
  return props.theme.primaryBtnBgd;
}, function (props) {
  return props.theme.textColorLT;
}, function (props) {
  return props.theme.textColor;
});

var propTypes = {
  datasets: _propTypes2.default.object.isRequired,
  selectedDataset: _propTypes2.default.string,
  dataType: _propTypes2.default.string.isRequired,
  filtered: _propTypes2.default.bool.isRequired,
  // callbacks
  onClose: _propTypes2.default.func.isRequired,
  onChangeExportSelectedDataset: _propTypes2.default.func.isRequired,
  onChangeExportDataType: _propTypes2.default.func.isRequired,
  onChangeExportFiltered: _propTypes2.default.func.isRequired,
  onChangeExportConfig: _propTypes2.default.func.isRequired
};

var getDataRowCount = function getDataRowCount(datasets, selectedDataset, filtered) {
  var selectedData = datasets[selectedDataset];
  if (!selectedData) {
    return Object.keys(datasets).length + ' Files ';
  }
  var allData = selectedData.allData,
      data = selectedData.data;

  var rowCount = filtered ? data.length : allData.length;
  return rowCount.toLocaleString('en') + ' Rows';
};

var ExportDataModal = function ExportDataModal(_ref) {
  var datasets = _ref.datasets,
      selectedDataset = _ref.selectedDataset,
      dataType = _ref.dataType,
      filtered = _ref.filtered,
      config = _ref.config,
      onChangeExportDataType = _ref.onChangeExportDataType,
      onChangeExportSelectedDataset = _ref.onChangeExportSelectedDataset,
      onChangeExportFiltered = _ref.onChangeExportFiltered;
  return _react2.default.createElement(
    'div',
    { className: 'export-data-modal' },
    _react2.default.createElement(
      _styledComponents3.StyledModalContent,
      null,
      _react2.default.createElement(
        'div',
        null,
        _react2.default.createElement(
          StyledExportDataSection,
          null,
          _react2.default.createElement(
            'div',
            { className: 'description' },
            _react2.default.createElement(
              'div',
              { className: 'title' },
              'Dataset'
            ),
            _react2.default.createElement(
              'div',
              { className: 'subtitle' },
              'Choose the datasets you want to export'
            )
          ),
          _react2.default.createElement(
            'div',
            { className: 'selection' },
            _react2.default.createElement(
              'select',
              { value: selectedDataset, onChange: function onChange(e) {
                  return onChangeExportSelectedDataset({ dataset: e.target.value });
                } },
              ['All'].concat(Object.keys(datasets)).map(function (d) {
                return _react2.default.createElement(
                  'option',
                  { key: d, value: d },
                  datasets[d] && datasets[d].label || d
                );
              })
            )
          )
        ),
        _react2.default.createElement(
          StyledExportDataSection,
          null,
          _react2.default.createElement(
            'div',
            { className: 'description' },
            _react2.default.createElement(
              'div',
              { className: 'title' },
              'Data Type'
            ),
            _react2.default.createElement(
              'div',
              { className: 'subtitle' },
              'Choose the type of data you want to export'
            )
          ),
          _react2.default.createElement(
            'div',
            { className: 'selection' },
            _defaultSettings.EXPORT_DATA_TYPE_OPTIONS.map(function (op) {
              return _react2.default.createElement(
                StyledDataType,
                {
                  key: op.id,
                  selected: dataType === op.id,
                  available: op.available,
                  onClick: function onClick() {
                    return op.available && onChangeExportDataType({ dataType: op.id });
                  }
                },
                _react2.default.createElement(_icons.FileType, { ext: op.label, height: '80px', fontSize: '11px' })
              );
            })
          )
        ),
        _react2.default.createElement(
          StyledExportDataSection,
          null,
          _react2.default.createElement(
            'div',
            { className: 'description' },
            _react2.default.createElement(
              'div',
              { className: 'title' },
              'Filter Data'
            ),
            _react2.default.createElement(
              'div',
              { className: 'subtitle' },
              'You can choose exporting original data or filtered data'
            )
          ),
          _react2.default.createElement(
            'div',
            { className: 'selection' },
            _react2.default.createElement(
              StyledFilteredDataOption,
              { selected: !filtered, onClick: function onClick() {
                  return onChangeExportFiltered({ filtered: false });
                } },
              _react2.default.createElement(
                'div',
                { className: 'filtered-title' },
                'Unfiltered Data'
              ),
              _react2.default.createElement(
                'div',
                { className: 'filtered-subtitle' },
                getDataRowCount(datasets, selectedDataset, false)
              )
            ),
            _react2.default.createElement(
              StyledFilteredDataOption,
              { selected: filtered, onClick: function onClick() {
                  return onChangeExportFiltered({ filtered: true });
                } },
              _react2.default.createElement(
                'div',
                { className: 'filtered-title' },
                'Filtered Data'
              ),
              _react2.default.createElement(
                'div',
                { className: 'filtered-subtitle' },
                getDataRowCount(datasets, selectedDataset, true)
              )
            )
          )
        )
      )
    )
  );
};

ExportDataModal.propTypes = propTypes;

var ExportDataModalFactory = function ExportDataModalFactory() {
  return ExportDataModal;
};
exports.default = ExportDataModalFactory;
